{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. annotate MNV using VEP\n",
    "\n",
    "In this section we will walk through how to annotate MNV using vep in hail.\n",
    "\n",
    "In order to reproduce this tutorial results, you will need to set up your vep configuration. \n",
    "See [Hail docs](https://hail.is/docs/0.2/methods/genetics.html?highlight=vep#hail.methods.vep) for vep configuration set up.\n",
    "\n",
    "Similar to the last tutorial, we ran\n",
    "\n",
    "`cluster start mycluster --num-preemptible-workers 8 --vep --init gs://gnomad-public/tools/inits/master-init.sh`\n",
    "Followed by\n",
    "`cluster connect mycluster notebook`\n",
    "\n",
    "To build the jupyter notebook in google cloud cluster.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing hail, and specifying the vep config\n",
    "import hail as hl\n",
    "import hail.expr.aggregators as agg\n",
    "from typing import *\n",
    "vep_config = \"gs://gnomad-resources/loftee-beta/vep85-loftee-gcloud.json\"  # this is the config that actually works!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Initializing Spark and Hail with default parameters...\n",
      "Running on Apache Spark version 2.2.3\n",
      "SparkUI available at http://10.128.0.44:4043\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.11-61a2083d5773\n",
      "LOGGING: writing to /home/hail/hail-20190310-0052-0.2.11-61a2083d5773.log\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+------------+---------------+--------------+-------+----------+-------+\n",
      "| locus         | alleles    | prev_locus    | prev_alleles |  dist |       AF |    AC |\n",
      "+---------------+------------+---------------+--------------+-------+----------+-------+\n",
      "| locus<GRCh37> | array<str> | locus<GRCh37> | array<str>   | int32 |  float64 | int32 |\n",
      "+---------------+------------+---------------+--------------+-------+----------+-------+\n",
      "| 1:10506       | [\"C\",\"G\"]  | 1:10505       | [\"A\",\"T\"]    |     1 | 2.00e-04 |     1 |\n",
      "| 1:13118       | [\"A\",\"G\"]  | 1:13116       | [\"T\",\"G\"]    |     2 | 9.70e-02 |   486 |\n",
      "| 1:48328       | [\"A\",\"T\"]  | 1:48327       | [\"C\",\"A\"]    |     1 | 4.39e-03 |    22 |\n",
      "| 1:49989       | [\"A\",\"G\"]  | 1:49988       | [\"T\",\"A\"]    |     1 | 5.79e-03 |    29 |\n",
      "| 1:51049       | [\"A\",\"C\"]  | 1:51047       | [\"A\",\"T\"]    |     2 | 1.60e-03 |     8 |\n",
      "| 1:51050       | [\"A\",\"T\"]  | 1:51049       | [\"A\",\"C\"]    |     1 | 1.60e-03 |     8 |\n",
      "| 1:55165       | [\"A\",\"G\"]  | 1:55164       | [\"C\",\"A\"]    |     1 | 2.00e-04 |     1 |\n",
      "| 1:55545       | [\"C\",\"T\"]  | 1:55544       | [\"T\",\"G\"]    |     1 | 2.39e-01 |  1198 |\n",
      "| 1:57264       | [\"T\",\"G\"]  | 1:57262       | [\"T\",\"G\"]    |     2 | 1.40e-03 |     7 |\n",
      "| 1:74792       | [\"G\",\"A\"]  | 1:74790       | [\"C\",\"G\"]    |     2 | 3.77e-02 |   189 |\n",
      "+---------------+------------+---------------+--------------+-------+----------+-------+\n",
      "\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| filters  |  prev_AF | prev_AC | prev_filters | n_hethet | n_hethom | n_homhom |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| set<str> |  float64 |   int32 | set<str>     |    int64 |    int64 |    int64 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| {}       | 2.00e-04 |       1 | {}           |        1 |        0 |        0 |\n",
      "| {}       | 9.70e-02 |     486 | {}           |      414 |       36 |        0 |\n",
      "| {}       | 4.39e-03 |      22 | {}           |       22 |        0 |        0 |\n",
      "| {}       | 5.79e-03 |      29 | {}           |       29 |        0 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 9.23e-01 |    4624 | {}           |        0 |        0 |        1 |\n",
      "| {}       | 2.00e-04 |       1 | {}           |        1 |        0 |        0 |\n",
      "| {}       | 1.40e-03 |       7 | {}           |        7 |        0 |        0 |\n",
      "| {}       | 3.77e-02 |     189 | {}           |      187 |        1 |        0 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "\n",
      "+---------+\n",
      "| n_total |\n",
      "+---------+\n",
      "|   int64 |\n",
      "+---------+\n",
      "|       1 |\n",
      "|     450 |\n",
      "|      22 |\n",
      "|      29 |\n",
      "|       8 |\n",
      "|       8 |\n",
      "|       1 |\n",
      "|       1 |\n",
      "|       7 |\n",
      "|     188 |\n",
      "+---------+\n",
      "showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#import the mnv table created in the tutorial step 1.\n",
    "mnv = hl.read_table(\"gs://gnomad-qingbowang/MNV/demo_et_combined.ht\")\n",
    "mnv.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## annotating the SNP effects\n",
    "\n",
    "This is relatively straightforward. Let the SNP of `prev_locus` be SNP1, and the other one be SNP2 (i.e. SNP2 is downstream).\n",
    "\n",
    "Since the vep function in hail takes `locus` and `alleles`, it is easier if we first annotate the SNP2 consequence by vep:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mnv = mnv.key_by(\"locus\",\"alleles\")\n",
    "mnv = hl.vep(mnv, vep_config, name=\"snp2_vep\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we want to annotate the SNP1 consequence using vep. In order to do this, we need to change the row name a little bit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:00:19 Hail: INFO: Coerced sorted dataset\n"
     ]
    }
   ],
   "source": [
    "mnv = mnv.key_by() #unkey first\n",
    "mnv = mnv.rename({'locus' : 'snp2_locus', 'alleles' : 'snp2_alleles',\"prev_locus\":\"locus\", \"prev_alleles\":\"alleles\"}) #rename the snp1 locus and alleles as locus, allele\n",
    "mnv = mnv.key_by('locus', 'alleles') #and re-key\n",
    "mnv = hl.vep(mnv, vep_config, name=\"snp1_vep\") #and vep\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's see that the individual effects are correctly annotated by vep:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:01:03 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| snp2_locus    | snp2_alleles | locus         | alleles    |  dist |       AF |    AC |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| locus<GRCh37> | array<str>   | locus<GRCh37> | array<str> | int32 |  float64 | int32 |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| 1:10506       | [\"C\",\"G\"]    | 1:10505       | [\"A\",\"T\"]  |     1 | 2.00e-04 |     1 |\n",
      "| 1:13118       | [\"A\",\"G\"]    | 1:13116       | [\"T\",\"G\"]  |     2 | 9.70e-02 |   486 |\n",
      "| 1:48328       | [\"A\",\"T\"]    | 1:48327       | [\"C\",\"A\"]  |     1 | 4.39e-03 |    22 |\n",
      "| 1:49989       | [\"A\",\"G\"]    | 1:49988       | [\"T\",\"A\"]  |     1 | 5.79e-03 |    29 |\n",
      "| 1:51049       | [\"A\",\"C\"]    | 1:51047       | [\"A\",\"T\"]  |     2 | 1.60e-03 |     8 |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| filters  |  prev_AF | prev_AC | prev_filters | n_hethet | n_hethom | n_homhom |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| set<str> |  float64 |   int32 | set<str>     |    int64 |    int64 |    int64 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| {}       | 2.00e-04 |       1 | {}           |        1 |        0 |        0 |\n",
      "| {}       | 9.70e-02 |     486 | {}           |      414 |       36 |        0 |\n",
      "| {}       | 4.39e-03 |      22 | {}           |       22 |        0 |        0 |\n",
      "| {}       | 5.79e-03 |      29 | {}           |       29 |        0 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "| n_total | snp2_vep.assembly_name | snp2_vep.allele_string | snp2_vep.ancestral |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|   int64 | str                    | str                    | str                |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|       1 | \"GRCh37\"               | \"C/G\"                  | NA                 |\n",
      "|     450 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "|      22 | \"GRCh37\"               | \"A/T\"                  | NA                 |\n",
      "|      29 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"A/C\"                  | NA                 |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",0.00e+00,NA,NA,\"C/G\",\"G\",8.00e-04,NA,10506,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"G\",1.53e-01,NA,NA,\"A/G\",\"G\",2.95e-02,NA,13118,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",1.36e-02,NA,48328,\"T\",1.40e-03,NA,NA,\"T\",3.00e-03,NA... |\n",
      "| [(\"G\",1.00e-03,NA,NA,\"A/G\",\"G\",1.74e-02,NA,49989,\"G\",1.40e-03,NA,NA,\"G\",4.00e-03,NA... |\n",
      "| [(\"C\",0.00e+00,NA,NA,\"A/C\",\"C\",0.00e+00,NA,51049,\"C\",0.00e+00,NA,NA,\"C\",3.00e-03,NA... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| snp2_vep.context | snp2_vep.end | snp2_vep.id   | snp2_vep.input                |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| str              |        int32 | str           | str                           |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| NA               |        10506 | \"1_10506_C/G\" | \"1\\t10506\\t.\\tC\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        13118 | \"1_13118_A/G\" | \"1\\t13118\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        48328 | \"1_48328_A/T\" | \"1\\t48328\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |        49989 | \"1_49989_A/G\" | \"1\\t49989\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        51049 | \"1_51049_A/C\" | \"1\\t51049\\t.\\tA\\tC\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------+\n",
      "| snp2_vep.most_severe_consequence |\n",
      "+----------------------------------+\n",
      "| str                              |\n",
      "+----------------------------------+\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"intron_variant\"                 |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "+----------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp2_vep.seq_region_name | snp2_vep.start | snp2_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          10506 |               1 |\n",
      "| \"1\"                      |          13118 |               1 |\n",
      "| \"1\"                      |          48328 |               1 |\n",
      "| \"1\"                      |          49989 |               1 |\n",
      "| \"1\"                      |          51049 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],38... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4721... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],3060... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2000... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+------------------------+------------------------+\n",
      "| snp2_vep.variant_class | snp1_vep.assembly_name | snp1_vep.allele_string |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| str                    | str                    | str                    |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/T\"                  |\n",
      "+------------------------+------------------------+------------------------+\n",
      "\n",
      "+--------------------+\n",
      "| snp1_vep.ancestral |\n",
      "+--------------------+\n",
      "| str                |\n",
      "+--------------------+\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",8.00e-04,NA,10505,\"T\",0.00e+00,NA,NA,\"T\",0.00e+00,NA... |\n",
      "| [(\"G\",1.53e-01,NA,NA,\"T/G\",\"G\",2.95e-02,NA,13116,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"C/A\",\"A\",1.36e-02,NA,48327,\"A\",1.40e-03,NA,NA,\"A\",3.00e-03,NA... |\n",
      "| [(\"A\",1.00e-03,NA,NA,\"T/A\",\"A\",1.74e-02,NA,49988,\"A\",1.40e-03,NA,NA,\"A\",4.00e-03,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",0.00e+00,NA,51047,\"T\",0.00e+00,NA,NA,\"T\",3.00e-03,NA... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| snp1_vep.context | snp1_vep.end | snp1_vep.id   | snp1_vep.input                |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| str              |        int32 | str           | str                           |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| NA               |        10505 | \"1_10505_A/T\" | \"1\\t10505\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |        13116 | \"1_13116_T/G\" | \"1\\t13116\\t.\\tT\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        48327 | \"1_48327_C/A\" | \"1\\t48327\\t.\\tC\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |        49988 | \"1_49988_T/A\" | \"1\\t49988\\t.\\tT\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |        51047 | \"1_51047_A/T\" | \"1\\t51047\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------+\n",
      "| snp1_vep.most_severe_consequence |\n",
      "+----------------------------------+\n",
      "| str                              |\n",
      "+----------------------------------+\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"intron_variant\"                 |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "+----------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp1_vep.seq_region_name | snp1_vep.start | snp1_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          10505 |               1 |\n",
      "| \"1\"                      |          13116 |               1 |\n",
      "| \"1\"                      |          48327 |               1 |\n",
      "| \"1\"                      |          49988 |               1 |\n",
      "| \"1\"                      |          51047 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],38... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4722... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],3061... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2002... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+\n",
      "| snp1_vep.variant_class |\n",
      "+------------------------+\n",
      "| str                    |\n",
      "+------------------------+\n",
      "| \"SNV\"                  |\n",
      "| \"SNV\"                  |\n",
      "| \"SNV\"                  |\n",
      "| \"SNV\"                  |\n",
      "| \"SNV\"                  |\n",
      "+------------------------+\n",
      "showing top 5 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mnv.show(5) #check"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## annotating the MNV effects\n",
    "\n",
    "This is still relatively straightforward, when the SNPs are next to each other, so let's focus on such MNVs first and do the annotation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:07:34 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:07:34 Hail: INFO: Coerced sorted dataset\n"
     ]
    }
   ],
   "source": [
    "#let t1 be those with distance==1\n",
    "t1 = mnv.filter(mnv.dist==1)\n",
    "t1 = t1.annotate(refs=t1.alleles[0] +t1.snp2_alleles[0], alts=t1.alleles[1] +t1.snp2_alleles[1])# annotate combined refs, and alts\n",
    "t1 = t1.annotate(mnv_alleles = [t1.refs,t1.alts]) #and let it be mnv_alleles\n",
    "t1 = t1.key_by()#to re-key\n",
    "t1 = t1.rename({'alleles' : 'snp1_alleles',\"mnv_alleles\":\"alleles\"}) #now the alleles is that of mnv (note. locus does not change. same as snp1)\n",
    "t1 = t1.key_by('locus', 'alleles') #and re-key\n",
    "t1 = hl.vep(t1, vep_config, name=\"mnv_vep\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:07:44 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:07:49 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| snp2_locus    | snp2_alleles | locus         | snp1_alleles |  dist |       AF |    AC |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| locus<GRCh37> | array<str>   | locus<GRCh37> | array<str>   | int32 |  float64 | int32 |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| 1:10506       | [\"C\",\"G\"]    | 1:10505       | [\"A\",\"T\"]    |     1 | 2.00e-04 |     1 |\n",
      "| 1:48328       | [\"A\",\"T\"]    | 1:48327       | [\"C\",\"A\"]    |     1 | 4.39e-03 |    22 |\n",
      "| 1:49989       | [\"A\",\"G\"]    | 1:49988       | [\"T\",\"A\"]    |     1 | 5.79e-03 |    29 |\n",
      "| 1:51050       | [\"A\",\"T\"]    | 1:51049       | [\"A\",\"C\"]    |     1 | 1.60e-03 |     8 |\n",
      "| 1:55165       | [\"A\",\"G\"]    | 1:55164       | [\"C\",\"A\"]    |     1 | 2.00e-04 |     1 |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| filters  |  prev_AF | prev_AC | prev_filters | n_hethet | n_hethom | n_homhom |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| set<str> |  float64 |   int32 | set<str>     |    int64 |    int64 |    int64 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| {}       | 2.00e-04 |       1 | {}           |        1 |        0 |        0 |\n",
      "| {}       | 4.39e-03 |      22 | {}           |       22 |        0 |        0 |\n",
      "| {}       | 5.79e-03 |      29 | {}           |       29 |        0 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 9.23e-01 |    4624 | {}           |        0 |        0 |        1 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "| n_total | snp2_vep.assembly_name | snp2_vep.allele_string | snp2_vep.ancestral |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|   int64 | str                    | str                    | str                |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|       1 | \"GRCh37\"               | \"C/G\"                  | NA                 |\n",
      "|      22 | \"GRCh37\"               | \"A/T\"                  | NA                 |\n",
      "|      29 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"A/T\"                  | NA                 |\n",
      "|       1 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",0.00e+00,NA,NA,\"C/G\",\"G\",8.00e-04,NA,10506,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",1.36e-02,NA,48328,\"T\",1.40e-03,NA,NA,\"T\",3.00e-03,NA... |\n",
      "| [(\"G\",1.00e-03,NA,NA,\"A/G\",\"G\",1.74e-02,NA,49989,\"G\",1.40e-03,NA,NA,\"G\",4.00e-03,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",0.00e+00,NA,51050,\"T\",0.00e+00,NA,NA,\"T\",3.00e-03,NA... |\n",
      "| [(\"G\",0.00e+00,NA,NA,\"A/G\",\"G\",8.00e-04,NA,55165,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| snp2_vep.context | snp2_vep.end | snp2_vep.id   | snp2_vep.input                |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| str              |        int32 | str           | str                           |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| NA               |        10506 | \"1_10506_C/G\" | \"1\\t10506\\t.\\tC\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        48328 | \"1_48328_A/T\" | \"1\\t48328\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |        49989 | \"1_49989_A/G\" | \"1\\t49989\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |        51050 | \"1_51050_A/T\" | \"1\\t51050\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |        55165 | \"1_55165_A/G\" | \"1\\t55165\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------+\n",
      "| snp2_vep.most_severe_consequence |\n",
      "+----------------------------------+\n",
      "| str                              |\n",
      "+----------------------------------+\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"downstream_gene_variant\"        |\n",
      "+----------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp2_vep.seq_region_name | snp2_vep.start | snp2_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          10506 |               1 |\n",
      "| \"1\"                      |          48328 |               1 |\n",
      "| \"1\"                      |          49989 |               1 |\n",
      "| \"1\"                      |          51050 |               1 |\n",
      "| \"1\"                      |          55165 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],38... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4721... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],3060... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1999... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],22... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+------------------------+------------------------+\n",
      "| snp2_vep.variant_class | snp1_vep.assembly_name | snp1_vep.allele_string |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| str                    | str                    | str                    |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/C\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/A\"                  |\n",
      "+------------------------+------------------------+------------------------+\n",
      "\n",
      "+--------------------+\n",
      "| snp1_vep.ancestral |\n",
      "+--------------------+\n",
      "| str                |\n",
      "+--------------------+\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",8.00e-04,NA,10505,\"T\",0.00e+00,NA,NA,\"T\",0.00e+00,NA... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"C/A\",\"A\",1.36e-02,NA,48327,\"A\",1.40e-03,NA,NA,\"A\",3.00e-03,NA... |\n",
      "| [(\"A\",1.00e-03,NA,NA,\"T/A\",\"A\",1.74e-02,NA,49988,\"A\",1.40e-03,NA,NA,\"A\",4.00e-03,NA... |\n",
      "| [(\"C\",0.00e+00,NA,NA,\"A/C\",\"C\",0.00e+00,NA,51049,\"C\",0.00e+00,NA,NA,\"C\",3.00e-03,NA... |\n",
      "| [(\"A\",9.86e-01,NA,NA,\"C/A\",\"A\",7.43e-01,NA,55164,\"A\",9.64e-01,NA,NA,\"A\",1.00e+00,NA... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| snp1_vep.context | snp1_vep.end | snp1_vep.id   | snp1_vep.input                |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| str              |        int32 | str           | str                           |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "| NA               |        10505 | \"1_10505_A/T\" | \"1\\t10505\\t.\\tA\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |        48327 | \"1_48327_C/A\" | \"1\\t48327\\t.\\tC\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |        49988 | \"1_49988_T/A\" | \"1\\t49988\\t.\\tT\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |        51049 | \"1_51049_A/C\" | \"1\\t51049\\t.\\tA\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |        55164 | \"1_55164_C/A\" | \"1\\t55164\\t.\\tC\\tA\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+---------------+-------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------+\n",
      "| snp1_vep.most_severe_consequence |\n",
      "+----------------------------------+\n",
      "| str                              |\n",
      "+----------------------------------+\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"upstream_gene_variant\"          |\n",
      "| \"downstream_gene_variant\"        |\n",
      "+----------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp1_vep.seq_region_name | snp1_vep.start | snp1_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          10505 |               1 |\n",
      "| \"1\"                      |          48327 |               1 |\n",
      "| \"1\"                      |          49988 |               1 |\n",
      "| \"1\"                      |          51049 |               1 |\n",
      "| \"1\"                      |          55164 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],38... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4722... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],3061... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2000... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],22... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+------+------+-------------+-----------------------+\n",
      "| snp1_vep.variant_class | refs | alts | alleles     | mnv_vep.assembly_name |\n",
      "+------------------------+------+------+-------------+-----------------------+\n",
      "| str                    | str  | str  | array<str>  | str                   |\n",
      "+------------------------+------+------+-------------+-----------------------+\n",
      "| \"SNV\"                  | \"AC\" | \"TG\" | [\"AC\",\"TG\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"CA\" | \"AT\" | [\"CA\",\"AT\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"TA\" | \"AG\" | [\"TA\",\"AG\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"AA\" | \"CT\" | [\"AA\",\"CT\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"CA\" | \"AG\" | [\"CA\",\"AG\"] | \"GRCh37\"              |\n",
      "+------------------------+------+------+-------------+-----------------------+\n",
      "\n",
      "+-----------------------+-------------------+\n",
      "| mnv_vep.allele_string | mnv_vep.ancestral |\n",
      "+-----------------------+-------------------+\n",
      "| str                   | str               |\n",
      "+-----------------------+-------------------+\n",
      "| \"AC/TG\"               | NA                |\n",
      "| \"CA/AT\"               | NA                |\n",
      "| \"TA/AG\"               | NA                |\n",
      "| \"AA/CT\"               | NA                |\n",
      "| \"CA/AG\"               | NA                |\n",
      "+-----------------------+-------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.colocated_variants                                                             |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-----------------+-------------+-----------------+---------------------------------+\n",
      "| mnv_vep.context | mnv_vep.end | mnv_vep.id      | mnv_vep.input                   |\n",
      "+-----------------+-------------+-----------------+---------------------------------+\n",
      "| str             |       int32 | str             | str                             |\n",
      "+-----------------+-------------+-----------------+---------------------------------+\n",
      "| NA              |       10506 | \"1_10505_AC/TG\" | \"1\\t10505\\t.\\tAC\\tTG\\t.\\t.\\tGT\" |\n",
      "| NA              |       48328 | \"1_48327_CA/AT\" | \"1\\t48327\\t.\\tCA\\tAT\\t.\\t.\\tGT\" |\n",
      "| NA              |       49989 | \"1_49988_TA/AG\" | \"1\\t49988\\t.\\tTA\\tAG\\t.\\t.\\tGT\" |\n",
      "| NA              |       51050 | \"1_51049_AA/CT\" | \"1\\t51049\\t.\\tAA\\tCT\\t.\\t.\\tGT\" |\n",
      "| NA              |       55165 | \"1_55164_CA/AG\" | \"1\\t55164\\t.\\tCA\\tAG\\t.\\t.\\tGT\" |\n",
      "+-----------------+-------------+-----------------+---------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.intergenic_consequences                                                        |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+---------------------------------+\n",
      "| mnv_vep.most_severe_consequence |\n",
      "+---------------------------------+\n",
      "| str                             |\n",
      "+---------------------------------+\n",
      "| \"upstream_gene_variant\"         |\n",
      "| \"upstream_gene_variant\"         |\n",
      "| \"upstream_gene_variant\"         |\n",
      "| \"upstream_gene_variant\"         |\n",
      "| \"downstream_gene_variant\"       |\n",
      "+---------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.motif_feature_consequences                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.regulatory_feature_consequences                                                |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-------------------------+---------------+----------------+\n",
      "| mnv_vep.seq_region_name | mnv_vep.start | mnv_vep.strand |\n",
      "+-------------------------+---------------+----------------+\n",
      "| str                     |         int32 |          int32 |\n",
      "+-------------------------+---------------+----------------+\n",
      "| \"1\"                     |         10505 |              1 |\n",
      "| \"1\"                     |         48327 |              1 |\n",
      "| \"1\"                     |         49988 |              1 |\n",
      "| \"1\"                     |         51049 |              1 |\n",
      "| \"1\"                     |         55164 |              1 |\n",
      "+-------------------------+---------------+----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.transcript_consequences                                                        |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],38... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4721... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],3060... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1999... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],22... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-----------------------+\n",
      "| mnv_vep.variant_class |\n",
      "+-----------------------+\n",
      "| str                   |\n",
      "+-----------------------+\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "+-----------------------+\n",
      "showing top 5 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t1.show(5) #check"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the SNPs are not to each other, it is a bit not straightforward because you need to fill the reference base pair that are in between, \n",
    "in order to pass to vep. \n",
    "Fortunately, our nice hail also allows us to do that. The first step is to load the reference genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "grch37 = hl.get_reference('GRCh37')\n",
    "grch37_fasta = 'gs://hail-common/references/human_g1k_v37.fasta.gz'\n",
    "grch37_fai = 'gs://hail-common/references/human_g1k_v37.fasta.fai'\n",
    "grch37.add_sequence(grch37_fasta, grch37_fai)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we can fill the context using `sequence_context`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "t2 = mnv.filter(mnv.dist==2)\n",
    "t2 = t2.annotate(refs = t2.alleles[0] + t2.locus.sequence_context(before=-1, after=1) + t2.snp2_alleles[0],\n",
    "                           alts = t2.alleles[1] + t2.locus.sequence_context(before=-1, after=1) + t2.snp2_alleles[1]) #fill the one reference sequence after the snp1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 02:01:48 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| snp2_locus    | snp2_alleles | locus         | alleles    |  dist |       AF |    AC |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| locus<GRCh37> | array<str>   | locus<GRCh37> | array<str> | int32 |  float64 | int32 |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "| 1:13118       | [\"A\",\"G\"]    | 1:13116       | [\"T\",\"G\"]  |     2 | 9.70e-02 |   486 |\n",
      "| 1:51049       | [\"A\",\"C\"]    | 1:51047       | [\"A\",\"T\"]  |     2 | 1.60e-03 |     8 |\n",
      "| 1:57264       | [\"T\",\"G\"]    | 1:57262       | [\"T\",\"G\"]  |     2 | 1.40e-03 |     7 |\n",
      "| 1:74792       | [\"G\",\"A\"]    | 1:74790       | [\"C\",\"G\"]  |     2 | 3.77e-02 |   189 |\n",
      "| 1:362907      | [\"G\",\"C\"]    | 1:362905      | [\"T\",\"G\"]  |     2 | 2.18e-02 |   109 |\n",
      "| 1:559985      | [\"C\",\"T\"]    | 1:559983      | [\"C\",\"T\"]  |     2 | 1.40e-02 |    70 |\n",
      "| 1:564862      | [\"T\",\"C\"]    | 1:564860      | [\"A\",\"G\"]  |     2 | 5.19e-02 |   260 |\n",
      "| 1:565006      | [\"T\",\"C\"]    | 1:565004      | [\"T\",\"A\"]  |     2 | 3.86e-01 |  1931 |\n",
      "| 1:565937      | [\"T\",\"C\"]    | 1:565935      | [\"C\",\"T\"]  |     2 | 3.93e-02 |   197 |\n",
      "| 1:568938      | [\"A\",\"G\"]    | 1:568936      | [\"G\",\"A\"]  |     2 | 2.20e-03 |    11 |\n",
      "+---------------+--------------+---------------+------------+-------+----------+-------+\n",
      "\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| filters  |  prev_AF | prev_AC | prev_filters | n_hethet | n_hethom | n_homhom |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| set<str> |  float64 |   int32 | set<str>     |    int64 |    int64 |    int64 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| {}       | 9.70e-02 |     486 | {}           |      414 |       36 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 1.40e-03 |       7 | {}           |        7 |        0 |        0 |\n",
      "| {}       | 3.77e-02 |     189 | {}           |      187 |        1 |        0 |\n",
      "| {}       | 1.26e-01 |     633 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 4.39e-03 |      22 | {}           |       17 |        0 |        1 |\n",
      "| {}       | 3.99e-04 |       2 | {}           |        0 |        0 |        1 |\n",
      "| {}       | 9.98e-04 |       5 | {}           |        0 |        1 |        1 |\n",
      "| {}       | 6.33e-02 |     317 | {}           |        3 |        0 |        0 |\n",
      "| {}       | 3.99e-03 |      20 | {}           |        2 |        3 |        0 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "| n_total | snp2_vep.assembly_name | snp2_vep.allele_string | snp2_vep.ancestral |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|   int64 | str                    | str                    | str                |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|     450 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"A/C\"                  | NA                 |\n",
      "|       7 | \"GRCh37\"               | \"T/G\"                  | NA                 |\n",
      "|     188 | \"GRCh37\"               | \"G/A\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"G/C\"                  | NA                 |\n",
      "|      18 | \"GRCh37\"               | \"C/T\"                  | NA                 |\n",
      "|       1 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       2 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       3 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       5 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",1.53e-01,NA,NA,\"A/G\",\"G\",2.95e-02,NA,13118,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"C\",0.00e+00,NA,NA,\"A/C\",\"C\",0.00e+00,NA,51049,\"C\",0.00e+00,NA,NA,\"C\",3.00e-03,NA... |\n",
      "| [(\"G\",7.20e-03,NA,NA,\"T/G\",\"G\",0.00e+00,NA,57264,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"A\",2.15e-02,NA,NA,\"G/A\",\"A\",7.56e-02,NA,74792,\"A\",4.90e-02,NA,NA,\"A\",8.90e-03,NA... |\n",
      "| [(\"C\",1.00e-03,NA,NA,\"G/C\",\"C\",7.79e-02,NA,362907,\"C\",4.30e-03,NA,NA,\"C\",2.00e-03,N... |\n",
      "| [(\"T\",1.23e-02,NA,NA,\"C/T\",\"T\",2.72e-02,NA,559985,\"T\",1.15e-02,NA,NA,\"T\",6.90e-03,N... |\n",
      "| [(\"C\",7.20e-03,NA,NA,\"T/C\",\"C\",1.23e-01,NA,564862,\"C\",1.30e-02,NA,NA,\"C\",3.67e-02,N... |\n",
      "| [(\"C\",2.25e-01,NA,NA,\"T/C\",\"C\",5.64e-01,NA,565006,\"C\",2.70e-01,NA,NA,\"C\",3.78e-01,N... |\n",
      "| [(\"C\",1.43e-02,NA,NA,\"T/C\",\"C\",8.77e-02,NA,565937,\"C\",7.20e-03,NA,NA,\"C\",3.77e-02,N... |\n",
      "| [(\"G\",2.00e-03,NA,NA,\"A/G\",\"G\",6.10e-03,NA,568938,\"G\",1.40e-03,NA,NA,\"G\",0.00e+00,N... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| snp2_vep.context | snp2_vep.end | snp2_vep.id    | snp2_vep.input                 |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| str              |        int32 | str            | str                            |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| NA               |        13118 | \"1_13118_A/G\"  | \"1\\t13118\\t.\\tA\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        51049 | \"1_51049_A/C\"  | \"1\\t51049\\t.\\tA\\tC\\t.\\t.\\tGT\"  |\n",
      "| NA               |        57264 | \"1_57264_T/G\"  | \"1\\t57264\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        74792 | \"1_74792_G/A\"  | \"1\\t74792\\t.\\tG\\tA\\t.\\t.\\tGT\"  |\n",
      "| NA               |       362907 | \"1_362907_G/C\" | \"1\\t362907\\t.\\tG\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       559985 | \"1_559985_C/T\" | \"1\\t559985\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       564862 | \"1_564862_T/C\" | \"1\\t564862\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       565006 | \"1_565006_T/C\" | \"1\\t565006\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       565937 | \"1_565937_T/C\" | \"1\\t565937\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       568938 | \"1_568938_A/G\" | \"1\\t568938\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------------------+\n",
      "| snp2_vep.most_severe_consequence     |\n",
      "+--------------------------------------+\n",
      "| str                                  |\n",
      "+--------------------------------------+\n",
      "| \"intron_variant\"                     |\n",
      "| \"upstream_gene_variant\"              |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "+--------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp2_vep.seq_region_name | snp2_vep.start | snp2_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          13118 |               1 |\n",
      "| \"1\"                      |          51049 |               1 |\n",
      "| \"1\"                      |          57264 |               1 |\n",
      "| \"1\"                      |          74792 |               1 |\n",
      "| \"1\"                      |         362907 |               1 |\n",
      "| \"1\"                      |         559985 |               1 |\n",
      "| \"1\"                      |         564862 |               1 |\n",
      "| \"1\"                      |         565006 |               1 |\n",
      "| \"1\"                      |         565937 |               1 |\n",
      "| \"1\"                      |         568938 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2000... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],23... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS30547.1\",NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS41220.1\",NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],47... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4457,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1592,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1448,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],517,N... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],942... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+------------------------+------------------------+\n",
      "| snp2_vep.variant_class | snp1_vep.assembly_name | snp1_vep.allele_string |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| str                    | str                    | str                    |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"G/A\"                  |\n",
      "+------------------------+------------------------+------------------------+\n",
      "\n",
      "+--------------------+\n",
      "| snp1_vep.ancestral |\n",
      "+--------------------+\n",
      "| str                |\n",
      "+--------------------+\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",1.53e-01,NA,NA,\"T/G\",\"G\",2.95e-02,NA,13116,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",0.00e+00,NA,51047,\"T\",0.00e+00,NA,NA,\"T\",3.00e-03,NA... |\n",
      "| [(\"G\",7.20e-03,NA,NA,\"T/G\",\"G\",0.00e+00,NA,57262,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"G\",2.15e-02,NA,NA,\"C/G\",\"G\",7.56e-02,NA,74790,\"G\",4.90e-02,NA,NA,\"G\",8.90e-03,NA... |\n",
      "| [(\"G\",4.60e-02,NA,NA,\"T/G\",\"G\",3.34e-01,NA,362905,\"G\",6.92e-02,NA,NA,\"G\",4.86e-02,N... |\n",
      "| [(\"T\",1.00e-03,NA,NA,\"C/T\",\"T\",1.13e-02,NA,559983,\"T\",8.60e-03,NA,NA,\"T\",0.00e+00,N... |\n",
      "| [(\"G\",0.00e+00,NA,NA,\"A/G\",\"G\",1.50e-03,NA,564860,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,N... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"T/A\",\"A\",3.80e-03,NA,565004,\"A\",0.00e+00,NA,NA,\"A\",0.00e+00,N... |\n",
      "| [(\"T\",4.29e-02,NA,NA,\"C/T\",\"T\",1.76e-01,NA,565935,\"T\",2.74e-02,NA,NA,\"T\",0.00e+00,N... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"G/A\",\"A\",1.44e-02,NA,568936,\"A\",1.40e-03,NA,NA,\"A\",0.00e+00,N... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| snp1_vep.context | snp1_vep.end | snp1_vep.id    | snp1_vep.input                 |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| str              |        int32 | str            | str                            |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| NA               |        13116 | \"1_13116_T/G\"  | \"1\\t13116\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        51047 | \"1_51047_A/T\"  | \"1\\t51047\\t.\\tA\\tT\\t.\\t.\\tGT\"  |\n",
      "| NA               |        57262 | \"1_57262_T/G\"  | \"1\\t57262\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        74790 | \"1_74790_C/G\"  | \"1\\t74790\\t.\\tC\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |       362905 | \"1_362905_T/G\" | \"1\\t362905\\t.\\tT\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |       559983 | \"1_559983_C/T\" | \"1\\t559983\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       564860 | \"1_564860_A/G\" | \"1\\t564860\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |       565004 | \"1_565004_T/A\" | \"1\\t565004\\t.\\tT\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |       565935 | \"1_565935_C/T\" | \"1\\t565935\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       568936 | \"1_568936_G/A\" | \"1\\t568936\\t.\\tG\\tA\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------------------+\n",
      "| snp1_vep.most_severe_consequence     |\n",
      "+--------------------------------------+\n",
      "| str                                  |\n",
      "+--------------------------------------+\n",
      "| \"intron_variant\"                     |\n",
      "| \"upstream_gene_variant\"              |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "+--------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp1_vep.seq_region_name | snp1_vep.start | snp1_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          13116 |               1 |\n",
      "| \"1\"                      |          51047 |               1 |\n",
      "| \"1\"                      |          57262 |               1 |\n",
      "| \"1\"                      |          74790 |               1 |\n",
      "| \"1\"                      |         362905 |               1 |\n",
      "| \"1\"                      |         559983 |               1 |\n",
      "| \"1\"                      |         564860 |               1 |\n",
      "| \"1\"                      |         565004 |               1 |\n",
      "| \"1\"                      |         565935 |               1 |\n",
      "| \"1\"                      |         568936 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2002... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],23... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS30547.1\",NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS41220.1\",NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],47... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4459,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1594,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1450,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],519,N... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],940... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+-------+-------+\n",
      "| snp1_vep.variant_class | refs  | alts  |\n",
      "+------------------------+-------+-------+\n",
      "| str                    | str   | str   |\n",
      "+------------------------+-------+-------+\n",
      "| \"SNV\"                  | \"TGA\" | \"GGG\" |\n",
      "| \"SNV\"                  | \"ATA\" | \"TTC\" |\n",
      "| \"SNV\"                  | \"TAT\" | \"GAG\" |\n",
      "| \"SNV\"                  | \"CTG\" | \"GTA\" |\n",
      "| \"SNV\"                  | \"TGG\" | \"GGC\" |\n",
      "| \"SNV\"                  | \"CGC\" | \"TGT\" |\n",
      "| \"SNV\"                  | \"AGT\" | \"GGC\" |\n",
      "| \"SNV\"                  | \"TAT\" | \"AAC\" |\n",
      "| \"SNV\"                  | \"CCT\" | \"TCC\" |\n",
      "| \"SNV\"                  | \"GTA\" | \"ATG\" |\n",
      "+------------------------+-------+-------+\n",
      "showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t2.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we are good to go!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 02:02:48 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 02:02:48 Hail: INFO: Coerced sorted dataset\n"
     ]
    }
   ],
   "source": [
    "t2 = t2.annotate(mnv_alleles = [t2.refs,t2.alts]) #and let it be mnv_alleles\n",
    "t2 = t2.key_by()#to re-key\n",
    "t2 = t2.rename({'alleles' : 'snp1_alleles',\"mnv_alleles\":\"alleles\"}) #now the alleles is that of mnv (note. locus does not change. same as snp1)\n",
    "t2 = t2.key_by('locus', 'alleles') #and re-key\n",
    "t2 = hl.vep(t2, vep_config, name=\"mnv_vep\", block_size=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 02:02:57 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 02:02:58 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| snp2_locus    | snp2_alleles | locus         | snp1_alleles |  dist |       AF |    AC |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| locus<GRCh37> | array<str>   | locus<GRCh37> | array<str>   | int32 |  float64 | int32 |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "| 1:13118       | [\"A\",\"G\"]    | 1:13116       | [\"T\",\"G\"]    |     2 | 9.70e-02 |   486 |\n",
      "| 1:51049       | [\"A\",\"C\"]    | 1:51047       | [\"A\",\"T\"]    |     2 | 1.60e-03 |     8 |\n",
      "| 1:57264       | [\"T\",\"G\"]    | 1:57262       | [\"T\",\"G\"]    |     2 | 1.40e-03 |     7 |\n",
      "| 1:74792       | [\"G\",\"A\"]    | 1:74790       | [\"C\",\"G\"]    |     2 | 3.77e-02 |   189 |\n",
      "| 1:362907      | [\"G\",\"C\"]    | 1:362905      | [\"T\",\"G\"]    |     2 | 2.18e-02 |   109 |\n",
      "| 1:559985      | [\"C\",\"T\"]    | 1:559983      | [\"C\",\"T\"]    |     2 | 1.40e-02 |    70 |\n",
      "| 1:564862      | [\"T\",\"C\"]    | 1:564860      | [\"A\",\"G\"]    |     2 | 5.19e-02 |   260 |\n",
      "| 1:565006      | [\"T\",\"C\"]    | 1:565004      | [\"T\",\"A\"]    |     2 | 3.86e-01 |  1931 |\n",
      "| 1:565937      | [\"T\",\"C\"]    | 1:565935      | [\"C\",\"T\"]    |     2 | 3.93e-02 |   197 |\n",
      "| 1:568938      | [\"A\",\"G\"]    | 1:568936      | [\"G\",\"A\"]    |     2 | 2.20e-03 |    11 |\n",
      "+---------------+--------------+---------------+--------------+-------+----------+-------+\n",
      "\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| filters  |  prev_AF | prev_AC | prev_filters | n_hethet | n_hethom | n_homhom |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| set<str> |  float64 |   int32 | set<str>     |    int64 |    int64 |    int64 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "| {}       | 9.70e-02 |     486 | {}           |      414 |       36 |        0 |\n",
      "| {}       | 1.60e-03 |       8 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 1.40e-03 |       7 | {}           |        7 |        0 |        0 |\n",
      "| {}       | 3.77e-02 |     189 | {}           |      187 |        1 |        0 |\n",
      "| {}       | 1.26e-01 |     633 | {}           |        8 |        0 |        0 |\n",
      "| {}       | 4.39e-03 |      22 | {}           |       17 |        0 |        1 |\n",
      "| {}       | 3.99e-04 |       2 | {}           |        0 |        0 |        1 |\n",
      "| {}       | 9.98e-04 |       5 | {}           |        0 |        1 |        1 |\n",
      "| {}       | 6.33e-02 |     317 | {}           |        3 |        0 |        0 |\n",
      "| {}       | 3.99e-03 |      20 | {}           |        2 |        3 |        0 |\n",
      "+----------+----------+---------+--------------+----------+----------+----------+\n",
      "\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "| n_total | snp2_vep.assembly_name | snp2_vep.allele_string | snp2_vep.ancestral |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|   int64 | str                    | str                    | str                |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "|     450 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"A/C\"                  | NA                 |\n",
      "|       7 | \"GRCh37\"               | \"T/G\"                  | NA                 |\n",
      "|     188 | \"GRCh37\"               | \"G/A\"                  | NA                 |\n",
      "|       8 | \"GRCh37\"               | \"G/C\"                  | NA                 |\n",
      "|      18 | \"GRCh37\"               | \"C/T\"                  | NA                 |\n",
      "|       1 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       2 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       3 | \"GRCh37\"               | \"T/C\"                  | NA                 |\n",
      "|       5 | \"GRCh37\"               | \"A/G\"                  | NA                 |\n",
      "+---------+------------------------+------------------------+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",1.53e-01,NA,NA,\"A/G\",\"G\",2.95e-02,NA,13118,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"C\",0.00e+00,NA,NA,\"A/C\",\"C\",0.00e+00,NA,51049,\"C\",0.00e+00,NA,NA,\"C\",3.00e-03,NA... |\n",
      "| [(\"G\",7.20e-03,NA,NA,\"T/G\",\"G\",0.00e+00,NA,57264,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"A\",2.15e-02,NA,NA,\"G/A\",\"A\",7.56e-02,NA,74792,\"A\",4.90e-02,NA,NA,\"A\",8.90e-03,NA... |\n",
      "| [(\"C\",1.00e-03,NA,NA,\"G/C\",\"C\",7.79e-02,NA,362907,\"C\",4.30e-03,NA,NA,\"C\",2.00e-03,N... |\n",
      "| [(\"T\",1.23e-02,NA,NA,\"C/T\",\"T\",2.72e-02,NA,559985,\"T\",1.15e-02,NA,NA,\"T\",6.90e-03,N... |\n",
      "| [(\"C\",7.20e-03,NA,NA,\"T/C\",\"C\",1.23e-01,NA,564862,\"C\",1.30e-02,NA,NA,\"C\",3.67e-02,N... |\n",
      "| [(\"C\",2.25e-01,NA,NA,\"T/C\",\"C\",5.64e-01,NA,565006,\"C\",2.70e-01,NA,NA,\"C\",3.78e-01,N... |\n",
      "| [(\"C\",1.43e-02,NA,NA,\"T/C\",\"C\",8.77e-02,NA,565937,\"C\",7.20e-03,NA,NA,\"C\",3.77e-02,N... |\n",
      "| [(\"G\",2.00e-03,NA,NA,\"A/G\",\"G\",6.10e-03,NA,568938,\"G\",1.40e-03,NA,NA,\"G\",0.00e+00,N... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| snp2_vep.context | snp2_vep.end | snp2_vep.id    | snp2_vep.input                 |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| str              |        int32 | str            | str                            |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| NA               |        13118 | \"1_13118_A/G\"  | \"1\\t13118\\t.\\tA\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        51049 | \"1_51049_A/C\"  | \"1\\t51049\\t.\\tA\\tC\\t.\\t.\\tGT\"  |\n",
      "| NA               |        57264 | \"1_57264_T/G\"  | \"1\\t57264\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        74792 | \"1_74792_G/A\"  | \"1\\t74792\\t.\\tG\\tA\\t.\\t.\\tGT\"  |\n",
      "| NA               |       362907 | \"1_362907_G/C\" | \"1\\t362907\\t.\\tG\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       559985 | \"1_559985_C/T\" | \"1\\t559985\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       564862 | \"1_564862_T/C\" | \"1\\t564862\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       565006 | \"1_565006_T/C\" | \"1\\t565006\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       565937 | \"1_565937_T/C\" | \"1\\t565937\\t.\\tT\\tC\\t.\\t.\\tGT\" |\n",
      "| NA               |       568938 | \"1_568938_A/G\" | \"1\\t568938\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------------------+\n",
      "| snp2_vep.most_severe_consequence     |\n",
      "+--------------------------------------+\n",
      "| str                                  |\n",
      "+--------------------------------------+\n",
      "| \"intron_variant\"                     |\n",
      "| \"upstream_gene_variant\"              |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "+--------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp2_vep.seq_region_name | snp2_vep.start | snp2_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          13118 |               1 |\n",
      "| \"1\"                      |          51049 |               1 |\n",
      "| \"1\"                      |          57264 |               1 |\n",
      "| \"1\"                      |          74792 |               1 |\n",
      "| \"1\"                      |         362907 |               1 |\n",
      "| \"1\"                      |         559985 |               1 |\n",
      "| \"1\"                      |         564862 |               1 |\n",
      "| \"1\"                      |         565006 |               1 |\n",
      "| \"1\"                      |         565937 |               1 |\n",
      "| \"1\"                      |         568938 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2000... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],23... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS30547.1\",NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS41220.1\",NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],47... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4457,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1592,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1448,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],517,N... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],942... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+------------------------+------------------------+\n",
      "| snp2_vep.variant_class | snp1_vep.assembly_name | snp1_vep.allele_string |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| str                    | str                    | str                    |\n",
      "+------------------------+------------------------+------------------------+\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"A/G\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"T/A\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"C/T\"                  |\n",
      "| \"SNV\"                  | \"GRCh37\"               | \"G/A\"                  |\n",
      "+------------------------+------------------------+------------------------+\n",
      "\n",
      "+--------------------+\n",
      "| snp1_vep.ancestral |\n",
      "+--------------------+\n",
      "| str                |\n",
      "+--------------------+\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "| NA                 |\n",
      "+--------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.colocated_variants                                                            |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(\"G\",1.53e-01,NA,NA,\"T/G\",\"G\",2.95e-02,NA,13116,\"G\",1.21e-01,NA,NA,\"G\",2.48e-02,NA... |\n",
      "| [(\"T\",0.00e+00,NA,NA,\"A/T\",\"T\",0.00e+00,NA,51047,\"T\",0.00e+00,NA,NA,\"T\",3.00e-03,NA... |\n",
      "| [(\"G\",7.20e-03,NA,NA,\"T/G\",\"G\",0.00e+00,NA,57262,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,NA... |\n",
      "| [(\"G\",2.15e-02,NA,NA,\"C/G\",\"G\",7.56e-02,NA,74790,\"G\",4.90e-02,NA,NA,\"G\",8.90e-03,NA... |\n",
      "| [(\"G\",4.60e-02,NA,NA,\"T/G\",\"G\",3.34e-01,NA,362905,\"G\",6.92e-02,NA,NA,\"G\",4.86e-02,N... |\n",
      "| [(\"T\",1.00e-03,NA,NA,\"C/T\",\"T\",1.13e-02,NA,559983,\"T\",8.60e-03,NA,NA,\"T\",0.00e+00,N... |\n",
      "| [(\"G\",0.00e+00,NA,NA,\"A/G\",\"G\",1.50e-03,NA,564860,\"G\",0.00e+00,NA,NA,\"G\",0.00e+00,N... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"T/A\",\"A\",3.80e-03,NA,565004,\"A\",0.00e+00,NA,NA,\"A\",0.00e+00,N... |\n",
      "| [(\"T\",4.29e-02,NA,NA,\"C/T\",\"T\",1.76e-01,NA,565935,\"T\",2.74e-02,NA,NA,\"T\",0.00e+00,N... |\n",
      "| [(\"A\",0.00e+00,NA,NA,\"G/A\",\"A\",1.44e-02,NA,568936,\"A\",1.40e-03,NA,NA,\"A\",0.00e+00,N... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| snp1_vep.context | snp1_vep.end | snp1_vep.id    | snp1_vep.input                 |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| str              |        int32 | str            | str                            |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "| NA               |        13116 | \"1_13116_T/G\"  | \"1\\t13116\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        51047 | \"1_51047_A/T\"  | \"1\\t51047\\t.\\tA\\tT\\t.\\t.\\tGT\"  |\n",
      "| NA               |        57262 | \"1_57262_T/G\"  | \"1\\t57262\\t.\\tT\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |        74790 | \"1_74790_C/G\"  | \"1\\t74790\\t.\\tC\\tG\\t.\\t.\\tGT\"  |\n",
      "| NA               |       362905 | \"1_362905_T/G\" | \"1\\t362905\\t.\\tT\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |       559983 | \"1_559983_C/T\" | \"1\\t559983\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       564860 | \"1_564860_A/G\" | \"1\\t564860\\t.\\tA\\tG\\t.\\t.\\tGT\" |\n",
      "| NA               |       565004 | \"1_565004_T/A\" | \"1\\t565004\\t.\\tT\\tA\\t.\\t.\\tGT\" |\n",
      "| NA               |       565935 | \"1_565935_C/T\" | \"1\\t565935\\t.\\tC\\tT\\t.\\t.\\tGT\" |\n",
      "| NA               |       568936 | \"1_568936_G/A\" | \"1\\t568936\\t.\\tG\\tA\\t.\\t.\\tGT\" |\n",
      "+------------------+--------------+----------------+--------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.intergenic_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------------------+\n",
      "| snp1_vep.most_severe_consequence     |\n",
      "+--------------------------------------+\n",
      "| str                                  |\n",
      "+--------------------------------------+\n",
      "| \"intron_variant\"                     |\n",
      "| \"upstream_gene_variant\"              |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "+--------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.motif_feature_consequences                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.regulatory_feature_consequences                                               |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------+----------------+-----------------+\n",
      "| snp1_vep.seq_region_name | snp1_vep.start | snp1_vep.strand |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| str                      |          int32 |           int32 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "| \"1\"                      |          13116 |               1 |\n",
      "| \"1\"                      |          51047 |               1 |\n",
      "| \"1\"                      |          57262 |               1 |\n",
      "| \"1\"                      |          74790 |               1 |\n",
      "| \"1\"                      |         362905 |               1 |\n",
      "| \"1\"                      |         559983 |               1 |\n",
      "| \"1\"                      |         564860 |               1 |\n",
      "| \"1\"                      |         565004 |               1 |\n",
      "| \"1\"                      |         565935 |               1 |\n",
      "| \"1\"                      |         568936 |               1 |\n",
      "+--------------------------+----------------+-----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp1_vep.transcript_consequences                                                       |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2002... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],23... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS30547.1\",NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS41220.1\",NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],47... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4459,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1594,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1450,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],519,N... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],940... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+------------------------+-------+-------+---------------+-----------------------+\n",
      "| snp1_vep.variant_class | refs  | alts  | alleles       | mnv_vep.assembly_name |\n",
      "+------------------------+-------+-------+---------------+-----------------------+\n",
      "| str                    | str   | str   | array<str>    | str                   |\n",
      "+------------------------+-------+-------+---------------+-----------------------+\n",
      "| \"SNV\"                  | \"TGA\" | \"GGG\" | [\"TGA\",\"GGG\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"ATA\" | \"TTC\" | [\"ATA\",\"TTC\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"TAT\" | \"GAG\" | [\"TAT\",\"GAG\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"CTG\" | \"GTA\" | [\"CTG\",\"GTA\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"TGG\" | \"GGC\" | [\"TGG\",\"GGC\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"CGC\" | \"TGT\" | [\"CGC\",\"TGT\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"AGT\" | \"GGC\" | [\"AGT\",\"GGC\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"TAT\" | \"AAC\" | [\"TAT\",\"AAC\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"CCT\" | \"TCC\" | [\"CCT\",\"TCC\"] | \"GRCh37\"              |\n",
      "| \"SNV\"                  | \"GTA\" | \"ATG\" | [\"GTA\",\"ATG\"] | \"GRCh37\"              |\n",
      "+------------------------+-------+-------+---------------+-----------------------+\n",
      "\n",
      "+-----------------------+-------------------+\n",
      "| mnv_vep.allele_string | mnv_vep.ancestral |\n",
      "+-----------------------+-------------------+\n",
      "| str                   | str               |\n",
      "+-----------------------+-------------------+\n",
      "| \"TGA/GGG\"             | NA                |\n",
      "| \"ATA/TTC\"             | NA                |\n",
      "| \"TAT/GAG\"             | NA                |\n",
      "| \"CTG/GTA\"             | NA                |\n",
      "| \"TGG/GGC\"             | NA                |\n",
      "| \"CGC/TGT\"             | NA                |\n",
      "| \"AGT/GGC\"             | NA                |\n",
      "| \"TAT/AAC\"             | NA                |\n",
      "| \"CCT/TCC\"             | NA                |\n",
      "| \"GTA/ATG\"             | NA                |\n",
      "+-----------------------+-------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.colocated_variants                                                             |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{aa_allele: str, aa_maf: float64, afr_allele: str, afr_maf: float64, al... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| [(NA,NA,NA,NA,\"TAT/AAC\",NA,NA,NA,565006,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,N... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-----------------+-------------+--------------------+\n",
      "| mnv_vep.context | mnv_vep.end | mnv_vep.id         |\n",
      "+-----------------+-------------+--------------------+\n",
      "| str             |       int32 | str                |\n",
      "+-----------------+-------------+--------------------+\n",
      "| NA              |       13118 | \"1_13116_TGA/GGG\"  |\n",
      "| NA              |       51049 | \"1_51047_ATA/TTC\"  |\n",
      "| NA              |       57264 | \"1_57262_TAT/GAG\"  |\n",
      "| NA              |       74792 | \"1_74790_CTG/GTA\"  |\n",
      "| NA              |      362907 | \"1_362905_TGG/GGC\" |\n",
      "| NA              |      559985 | \"1_559983_CGC/TGT\" |\n",
      "| NA              |      564862 | \"1_564860_AGT/GGC\" |\n",
      "| NA              |      565006 | \"1_565004_TAT/AAC\" |\n",
      "| NA              |      565937 | \"1_565935_CCT/TCC\" |\n",
      "| NA              |      568938 | \"1_568936_GTA/ATG\" |\n",
      "+-----------------+-------------+--------------------+\n",
      "\n",
      "+------------------------------------+\n",
      "| mnv_vep.input                      |\n",
      "+------------------------------------+\n",
      "| str                                |\n",
      "+------------------------------------+\n",
      "| \"1\\t13116\\t.\\tTGA\\tGGG\\t.\\t.\\tGT\"  |\n",
      "| \"1\\t51047\\t.\\tATA\\tTTC\\t.\\t.\\tGT\"  |\n",
      "| \"1\\t57262\\t.\\tTAT\\tGAG\\t.\\t.\\tGT\"  |\n",
      "| \"1\\t74790\\t.\\tCTG\\tGTA\\t.\\t.\\tGT\"  |\n",
      "| \"1\\t362905\\t.\\tTGG\\tGGC\\t.\\t.\\tGT\" |\n",
      "| \"1\\t559983\\t.\\tCGC\\tTGT\\t.\\t.\\tGT\" |\n",
      "| \"1\\t564860\\t.\\tAGT\\tGGC\\t.\\t.\\tGT\" |\n",
      "| \"1\\t565004\\t.\\tTAT\\tAAC\\t.\\t.\\tGT\" |\n",
      "| \"1\\t565935\\t.\\tCCT\\tTCC\\t.\\t.\\tGT\" |\n",
      "| \"1\\t568936\\t.\\tGTA\\tATG\\t.\\t.\\tGT\" |\n",
      "+------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.intergenic_consequences                                                        |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, impact: str, minimis... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+--------------------------------------+\n",
      "| mnv_vep.most_severe_consequence      |\n",
      "+--------------------------------------+\n",
      "| str                                  |\n",
      "+--------------------------------------+\n",
      "| \"intron_variant\"                     |\n",
      "| \"upstream_gene_variant\"              |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"downstream_gene_variant\"            |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"intron_variant\"                     |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "| \"non_coding_transcript_exon_variant\" |\n",
      "+--------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.motif_feature_consequences                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, consequence_terms: array<str>, high_inf_pos: str, i... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.regulatory_feature_consequences                                                |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, biotype: str, consequence_terms: array<str>, impact... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,\"CTCF_binding_site\",[\"regulatory_region_variant\"],\"MODIFIER\",1,\"ENSR00001576075... |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "| NA                                                                                     |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-------------------------+---------------+----------------+\n",
      "| mnv_vep.seq_region_name | mnv_vep.start | mnv_vep.strand |\n",
      "+-------------------------+---------------+----------------+\n",
      "| str                     |         int32 |          int32 |\n",
      "+-------------------------+---------------+----------------+\n",
      "| \"1\"                     |         13116 |              1 |\n",
      "| \"1\"                     |         51047 |              1 |\n",
      "| \"1\"                     |         57262 |              1 |\n",
      "| \"1\"                     |         74790 |              1 |\n",
      "| \"1\"                     |        362905 |              1 |\n",
      "| \"1\"                     |        559983 |              1 |\n",
      "| \"1\"                     |        564860 |              1 |\n",
      "| \"1\"                     |        565004 |              1 |\n",
      "| \"1\"                     |        565935 |              1 |\n",
      "| \"1\"                     |        568936 |              1 |\n",
      "+-------------------------+---------------+----------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| mnv_vep.transcript_consequences                                                        |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| array<struct{allele_num: int32, amino_acids: str, biotype: str, canonical: int32, c... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],12... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],2000... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",NA,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],23... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS30547.1\",NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],... |\n",
      "| [(1,NA,\"protein_coding\",1,\"CCDS41220.1\",NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],47... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],4457,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1592,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],1448,... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"upstream_gene_variant\"],517,N... |\n",
      "| [(1,NA,\"unprocessed_pseudogene\",1,NA,NA,NA,NA,NA,NA,[\"downstream_gene_variant\"],940... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+-----------------------+\n",
      "| mnv_vep.variant_class |\n",
      "+-----------------------+\n",
      "| str                   |\n",
      "+-----------------------+\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "| \"substitution\"        |\n",
      "+-----------------------+\n",
      "showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t2.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing the effects and categorizing MNV\n",
    "\n",
    "Now we have individual effect of SNP1, SNP2, as well as combined effect of MNV annotated. \n",
    "Next step is to compare those to categorize the MNVs. \n",
    "\n",
    "Are any \"gained\" nonsense in our dataset? How many MNVs are actually within a single codon reading frame?\n",
    "\n",
    "To answer these question, we will first clean the data and get per transcript annotation, focusing on canonical transcripts. \n",
    "\n",
    "(We will start from those with distance=1. Same logic can be applied to all the processes for distance 2 later.)\n",
    "\n",
    "(From here on, we are using some home made functions that are helpful for the analysis. \n",
    "Although there should be a more efficienc way to import those functions, here for the convenience we will copy and paste in the bottom of this notebook. \n",
    "See the \"functions\" cell for the functions that are used. In practice we need to define those functions first before running this process.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#first, select the essential columns (delete some)\n",
    "vepped_d1_essense = t1.key_by(\"locus\", \"refs\", \"alts\") \n",
    "vepped_d1_essense = vepped_d1_essense.select(\"AC\", \"prev_AC\", \"prev_AF\", \"n_hethet\", \"n_hethom\",\n",
    "                                                     \"n_homhom\", \"snp1_vep\", \"snp2_vep\", \"mnv_vep\")\n",
    "# filter to canonicals:\n",
    "canon_cons_d1 = filter_vep_to_canonical_transcripts(filter_vep_to_canonical_transcripts(\n",
    "filter_vep_to_canonical_transcripts(vepped_d1_essense, vep_root=\"snp1_vep\"), vep_root=\"snp2_vep\"),\n",
    "                                                            vep_root=\"mnv_vep\")\n",
    "\n",
    "###explode by consequence, since there might be more than 1 canonical transcript for a MNV\n",
    "canon_cons_d1 = canon_cons_d1.annotate(indices=hl.range(0, hl.min(\n",
    "            hl.len(canon_cons_d1.snp1_vep.transcript_consequences), hl.len(canon_cons_d1.snp2_vep.transcript_consequences))))\n",
    "canon_cons_d1 = canon_cons_d1.explode('indices')\n",
    "canon_cons_d1 = canon_cons_d1.transmute(\n",
    "            snp1_transcript_consequences=canon_cons_d1.snp1_vep.transcript_consequences[canon_cons_d1.indices],\n",
    "            snp2_transcript_consequences=canon_cons_d1.snp2_vep.transcript_consequences[canon_cons_d1.indices],\n",
    "            mnv_transcript_consequences=canon_cons_d1.mnv_vep.transcript_consequences[canon_cons_d1.indices],\n",
    "            )\n",
    "# keep necessary columns\n",
    "canon_cons_d1 = canon_cons_d1.select(\"AC\", \"prev_AC\", \"prev_AF\", \"n_hethet\", \"n_hethom\",\n",
    "                                             \"n_homhom\",\n",
    "                                             \"snp1_transcript_consequences\",\n",
    "                                             \"snp2_transcript_consequences\",\n",
    "                                             \"mnv_transcript_consequences\")\n",
    "\n",
    "# further annotate necessary ones\n",
    "# codons, consequence_terms, lof, lof_flags, transcript_id\n",
    "canon_cons_d1 = canon_cons_d1.annotate(\n",
    "            snp1_cons_term=canon_cons_d1.snp1_transcript_consequences.consequence_terms,\n",
    "            snp2_cons_term=canon_cons_d1.snp2_transcript_consequences.consequence_terms,\n",
    "            mnv_cons_term=canon_cons_d1.mnv_transcript_consequences.consequence_terms,\n",
    "            snp1_codons=canon_cons_d1.snp1_transcript_consequences.codons,\n",
    "            snp2_codons=canon_cons_d1.snp2_transcript_consequences.codons,\n",
    "            mnv_codons=canon_cons_d1.mnv_transcript_consequences.codons,\n",
    "            snp1_amino_acids=canon_cons_d1.snp1_transcript_consequences.amino_acids,\n",
    "            snp2_amino_acids=canon_cons_d1.snp2_transcript_consequences.amino_acids,\n",
    "            mnv_amino_acids=canon_cons_d1.mnv_transcript_consequences.amino_acids,\n",
    "            snp1_lof=canon_cons_d1.snp1_transcript_consequences.lof,\n",
    "            snp2_lof=canon_cons_d1.snp2_transcript_consequences.lof,\n",
    "            mnv_lof=canon_cons_d1.mnv_transcript_consequences.lof,\n",
    "            transcript_id=canon_cons_d1.snp1_transcript_consequences.transcript_id\n",
    "            )\n",
    "\n",
    "# filter to those that the codon are changed within a single reading frame\n",
    "canon_cons_d1 = canon_cons_d1.filter(\n",
    "            (canon_cons_d1.snp1_codons.length() == 7) & (canon_cons_d1.snp2_codons.length() == 7) & (\n",
    "            canon_cons_d1.mnv_codons.length() == 7))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In above we have filtered to select MNVs only falling on canonical transcripts, within a single codon reading frame.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:47:58 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:48:03 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:48:08 Hail: INFO: Coerced sorted dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+------+------+-------+---------+----------+----------+----------+\n",
      "| locus         | refs | alts |    AC | prev_AC |  prev_AF | n_hethet | n_hethom |\n",
      "+---------------+------+------+-------+---------+----------+----------+----------+\n",
      "| locus<GRCh37> | str  | str  | int32 |   int32 |  float64 |    int64 |    int64 |\n",
      "+---------------+------+------+-------+---------+----------+----------+----------+\n",
      "| 1:901922      | \"GC\" | \"AA\" |    53 |      53 | 1.06e-02 |       51 |        1 |\n",
      "| 1:909238      | \"GT\" | \"CC\" |     1 |    3889 | 7.77e-01 |        0 |        0 |\n",
      "+---------------+------+------+-------+---------+----------+----------+----------+\n",
      "\n",
      "+----------+-----------------------------------------+\n",
      "| n_homhom | snp1_transcript_consequences.allele_num |\n",
      "+----------+-----------------------------------------+\n",
      "|    int64 |                                   int32 |\n",
      "+----------+-----------------------------------------+\n",
      "|        0 |                                       1 |\n",
      "|        1 |                                       1 |\n",
      "+----------+-----------------------------------------+\n",
      "\n",
      "+------------------------------------------+--------------------------------------+\n",
      "| snp1_transcript_consequences.amino_acids | snp1_transcript_consequences.biotype |\n",
      "+------------------------------------------+--------------------------------------+\n",
      "| str                                      | str                                  |\n",
      "+------------------------------------------+--------------------------------------+\n",
      "| \"S/N\"                                    | \"protein_coding\"                     |\n",
      "| \"R/P\"                                    | \"protein_coding\"                     |\n",
      "+------------------------------------------+--------------------------------------+\n",
      "\n",
      "+----------------------------------------+-----------------------------------+\n",
      "| snp1_transcript_consequences.canonical | snp1_transcript_consequences.ccds |\n",
      "+----------------------------------------+-----------------------------------+\n",
      "|                                  int32 | str                               |\n",
      "+----------------------------------------+-----------------------------------+\n",
      "|                                      1 | \"CCDS4.1\"                         |\n",
      "|                                      1 | \"CCDS4.1\"                         |\n",
      "+----------------------------------------+-----------------------------------+\n",
      "\n",
      "+-----------------------------------------+---------------------------------------+\n",
      "| snp1_transcript_consequences.cdna_start | snp1_transcript_consequences.cdna_end |\n",
      "+-----------------------------------------+---------------------------------------+\n",
      "|                                   int32 |                                 int32 |\n",
      "+-----------------------------------------+---------------------------------------+\n",
      "|                                      46 |                                    46 |\n",
      "|                                    1495 |                                  1495 |\n",
      "+-----------------------------------------+---------------------------------------+\n",
      "\n",
      "+--------------------------------------+----------------------------------------+\n",
      "| snp1_transcript_consequences.cds_end | snp1_transcript_consequences.cds_start |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "|                                int32 |                                  int32 |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "|                                   11 |                                     11 |\n",
      "|                                 1460 |                                   1460 |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "\n",
      "+-------------------------------------+------------------------------------------------+\n",
      "| snp1_transcript_consequences.codons | snp1_transcript_consequences.consequence_terms |\n",
      "+-------------------------------------+------------------------------------------------+\n",
      "| str                                 | array<str>                                     |\n",
      "+-------------------------------------+------------------------------------------------+\n",
      "| \"aGc/aAc\"                           | [\"missense_variant\"]                           |\n",
      "| \"cGt/cCt\"                           | [\"missense_variant\"]                           |\n",
      "+-------------------------------------+------------------------------------------------+\n",
      "\n",
      "+---------------------------------------+\n",
      "| snp1_transcript_consequences.distance |\n",
      "+---------------------------------------+\n",
      "|                                 int32 |\n",
      "+---------------------------------------+\n",
      "|                                    NA |\n",
      "|                                    NA |\n",
      "+---------------------------------------+\n",
      "\n",
      "+---------------------------------------------------------+\n",
      "| snp1_transcript_consequences.domains                    |\n",
      "+---------------------------------------------------------+\n",
      "| array<struct{db: str, name: str}>                       |\n",
      "+---------------------------------------------------------+\n",
      "| NA                                                      |\n",
      "| [(\"hmmpanther\",\"PTHR22826\"),(\"hmmpanther\",\"PTHR22826\")] |\n",
      "+---------------------------------------------------------+\n",
      "\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| snp1_transcript_consequences.exon | snp1_transcript_consequences.gene_id |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| str                               | str                                  |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| \"1/16\"                            | \"ENSG00000187583\"                    |\n",
      "| \"14/16\"                           | \"ENSG00000187583\"                    |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "| snp1_transcript_consequences.gene_pheno | snp1_transcript_consequences.gene_symbol |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                   int32 | str                                      |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                      NA | \"PLEKHN1\"                                |\n",
      "|                                      NA | \"PLEKHN1\"                                |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| snp1_transcript_consequences.gene_symbol_source | snp1_transcript_consequences.hgnc_id |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| str                                             | str                                  |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| \"HGNC\"                                          | \"25284\"                              |\n",
      "| \"HGNC\"                                          | \"25284\"                              |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "\n",
      "+------------------------------------+------------------------------------+\n",
      "| snp1_transcript_consequences.hgvsc | snp1_transcript_consequences.hgvsp |\n",
      "+------------------------------------+------------------------------------+\n",
      "| str                                | str                                |\n",
      "+------------------------------------+------------------------------------+\n",
      "| \"ENST00000379410.3:c.11G>A\"        | \"ENSP00000368720.3:p.Ser4Asn\"      |\n",
      "| \"ENST00000379410.3:c.1460G>C\"      | \"ENSP00000368720.3:p.Arg487Pro\"    |\n",
      "+------------------------------------+------------------------------------+\n",
      "\n",
      "+------------------------------------------+-------------------------------------+\n",
      "| snp1_transcript_consequences.hgvs_offset | snp1_transcript_consequences.impact |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "|                                    int32 | str                                 |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "|                                       NA | \"MODERATE\"                          |\n",
      "|                                       NA | \"MODERATE\"                          |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "\n",
      "+-------------------------------------+----------------------------------+\n",
      "| snp1_transcript_consequences.intron | snp1_transcript_consequences.lof |\n",
      "+-------------------------------------+----------------------------------+\n",
      "| str                                 | str                              |\n",
      "+-------------------------------------+----------------------------------+\n",
      "| NA                                  | NA                               |\n",
      "| NA                                  | NA                               |\n",
      "+-------------------------------------+----------------------------------+\n",
      "\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| snp1_transcript_consequences.lof_flags | snp1_transcript_consequences.lof_filter |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| str                                    | str                                     |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| NA                                     | NA                                      |\n",
      "| NA                                     | NA                                      |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| snp1_transcript_consequences.lof_info | snp1_transcript_consequences.minimised |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| str                                   |                                  int32 |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| NA                                    |                                      1 |\n",
      "| NA                                    |                                      1 |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "\n",
      "+--------------------------------------------------+\n",
      "| snp1_transcript_consequences.polyphen_prediction |\n",
      "+--------------------------------------------------+\n",
      "| str                                              |\n",
      "+--------------------------------------------------+\n",
      "| \"benign\"                                         |\n",
      "| \"benign\"                                         |\n",
      "+--------------------------------------------------+\n",
      "\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "| snp1_transcript_consequences.polyphen_score | snp1_transcript_consequences.protein_end |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "|                                     float64 |                                    int32 |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "|                                    1.32e-01 |                                        4 |\n",
      "|                                    0.00e+00 |                                      487 |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "| snp1_transcript_consequences.protein_start | snp1_transcript_consequences.protein_id |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                      int32 | str                                     |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                          4 | \"ENSP00000368720\"                       |\n",
      "|                                        487 | \"ENSP00000368720\"                       |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| snp1_transcript_consequences.sift_prediction | snp1_transcript_consequences.sift_score |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| str                                          |                                 float64 |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| \"tolerated\"                                  |                                6.00e-02 |\n",
      "| \"tolerated\"                                  |                                3.10e-01 |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "\n",
      "+-------------------------------------+----------------------------------------+\n",
      "| snp1_transcript_consequences.strand | snp1_transcript_consequences.swissprot |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "|                               int32 | str                                    |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "|                                   1 | \"Q494U1\"                               |\n",
      "|                                   1 | \"Q494U1\"                               |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| snp1_transcript_consequences.transcript_id | snp1_transcript_consequences.trembl |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| str                                        | str                                 |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| \"ENST00000379410\"                          | \"J3KSM5\"                            |\n",
      "| \"ENST00000379410\"                          | \"J3KSM5\"                            |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| snp1_transcript_consequences.uniparc | snp1_transcript_consequences.variant_allele |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| str                                  | str                                         |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| \"UPI00001416D8\"                      | \"A\"                                         |\n",
      "| \"UPI00001416D8\"                      | \"C\"                                         |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "| snp2_transcript_consequences.allele_num | snp2_transcript_consequences.amino_acids |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                   int32 | str                                      |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                       1 | \"S/R\"                                    |\n",
      "|                                       1 | \"R\"                                      |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "\n",
      "+--------------------------------------+----------------------------------------+\n",
      "| snp2_transcript_consequences.biotype | snp2_transcript_consequences.canonical |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "| str                                  |                                  int32 |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "| \"protein_coding\"                     |                                      1 |\n",
      "| \"protein_coding\"                     |                                      1 |\n",
      "+--------------------------------------+----------------------------------------+\n",
      "\n",
      "+-----------------------------------+-----------------------------------------+\n",
      "| snp2_transcript_consequences.ccds | snp2_transcript_consequences.cdna_start |\n",
      "+-----------------------------------+-----------------------------------------+\n",
      "| str                               |                                   int32 |\n",
      "+-----------------------------------+-----------------------------------------+\n",
      "| \"CCDS4.1\"                         |                                      47 |\n",
      "| \"CCDS4.1\"                         |                                    1496 |\n",
      "+-----------------------------------+-----------------------------------------+\n",
      "\n",
      "+---------------------------------------+--------------------------------------+\n",
      "| snp2_transcript_consequences.cdna_end | snp2_transcript_consequences.cds_end |\n",
      "+---------------------------------------+--------------------------------------+\n",
      "|                                 int32 |                                int32 |\n",
      "+---------------------------------------+--------------------------------------+\n",
      "|                                    47 |                                   12 |\n",
      "|                                  1496 |                                 1461 |\n",
      "+---------------------------------------+--------------------------------------+\n",
      "\n",
      "+----------------------------------------+-------------------------------------+\n",
      "| snp2_transcript_consequences.cds_start | snp2_transcript_consequences.codons |\n",
      "+----------------------------------------+-------------------------------------+\n",
      "|                                  int32 | str                                 |\n",
      "+----------------------------------------+-------------------------------------+\n",
      "|                                     12 | \"agC/agA\"                           |\n",
      "|                                   1461 | \"cgT/cgC\"                           |\n",
      "+----------------------------------------+-------------------------------------+\n",
      "\n",
      "+------------------------------------------------+---------------------------------------+\n",
      "| snp2_transcript_consequences.consequence_terms | snp2_transcript_consequences.distance |\n",
      "+------------------------------------------------+---------------------------------------+\n",
      "| array<str>                                     |                                 int32 |\n",
      "+------------------------------------------------+---------------------------------------+\n",
      "| [\"missense_variant\"]                           |                                    NA |\n",
      "| [\"synonymous_variant\"]                         |                                    NA |\n",
      "+------------------------------------------------+---------------------------------------+\n",
      "\n",
      "+---------------------------------------------------------+\n",
      "| snp2_transcript_consequences.domains                    |\n",
      "+---------------------------------------------------------+\n",
      "| array<struct{db: str, name: str}>                       |\n",
      "+---------------------------------------------------------+\n",
      "| NA                                                      |\n",
      "| [(\"hmmpanther\",\"PTHR22826\"),(\"hmmpanther\",\"PTHR22826\")] |\n",
      "+---------------------------------------------------------+\n",
      "\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| snp2_transcript_consequences.exon | snp2_transcript_consequences.gene_id |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| str                               | str                                  |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "| \"1/16\"                            | \"ENSG00000187583\"                    |\n",
      "| \"14/16\"                           | \"ENSG00000187583\"                    |\n",
      "+-----------------------------------+--------------------------------------+\n",
      "\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "| snp2_transcript_consequences.gene_pheno | snp2_transcript_consequences.gene_symbol |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                   int32 | str                                      |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "|                                      NA | \"PLEKHN1\"                                |\n",
      "|                                      NA | \"PLEKHN1\"                                |\n",
      "+-----------------------------------------+------------------------------------------+\n",
      "\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| snp2_transcript_consequences.gene_symbol_source | snp2_transcript_consequences.hgnc_id |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| str                                             | str                                  |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "| \"HGNC\"                                          | \"25284\"                              |\n",
      "| \"HGNC\"                                          | \"25284\"                              |\n",
      "+-------------------------------------------------+--------------------------------------+\n",
      "\n",
      "+------------------------------------+------------------------------------+\n",
      "| snp2_transcript_consequences.hgvsc | snp2_transcript_consequences.hgvsp |\n",
      "+------------------------------------+------------------------------------+\n",
      "| str                                | str                                |\n",
      "+------------------------------------+------------------------------------+\n",
      "| \"ENST00000379410.3:c.12C>A\"        | \"ENSP00000368720.3:p.Ser4Arg\"      |\n",
      "| \"ENST00000379410.3:c.1461T>C\"      | \"ENST00000379410.3:c.1461T>C(p.=)\" |\n",
      "+------------------------------------+------------------------------------+\n",
      "\n",
      "+------------------------------------------+-------------------------------------+\n",
      "| snp2_transcript_consequences.hgvs_offset | snp2_transcript_consequences.impact |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "|                                    int32 | str                                 |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "|                                       NA | \"MODERATE\"                          |\n",
      "|                                       NA | \"LOW\"                               |\n",
      "+------------------------------------------+-------------------------------------+\n",
      "\n",
      "+-------------------------------------+----------------------------------+\n",
      "| snp2_transcript_consequences.intron | snp2_transcript_consequences.lof |\n",
      "+-------------------------------------+----------------------------------+\n",
      "| str                                 | str                              |\n",
      "+-------------------------------------+----------------------------------+\n",
      "| NA                                  | NA                               |\n",
      "| NA                                  | NA                               |\n",
      "+-------------------------------------+----------------------------------+\n",
      "\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| snp2_transcript_consequences.lof_flags | snp2_transcript_consequences.lof_filter |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| str                                    | str                                     |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| NA                                     | NA                                      |\n",
      "| NA                                     | NA                                      |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| snp2_transcript_consequences.lof_info                                                  |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| str                                                                                    |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| NA                                                                                     |\n",
      "| \"EXON_END:909431,DE_NOVO_DONOR_MES:-5.07903292316591,DE_NOVO_DONOR_MES_POS:-198,DE_... |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "\n",
      "+----------------------------------------+\n",
      "| snp2_transcript_consequences.minimised |\n",
      "+----------------------------------------+\n",
      "|                                  int32 |\n",
      "+----------------------------------------+\n",
      "|                                      1 |\n",
      "|                                      1 |\n",
      "+----------------------------------------+\n",
      "\n",
      "+--------------------------------------------------+\n",
      "| snp2_transcript_consequences.polyphen_prediction |\n",
      "+--------------------------------------------------+\n",
      "| str                                              |\n",
      "+--------------------------------------------------+\n",
      "| \"possibly_damaging\"                              |\n",
      "| NA                                               |\n",
      "+--------------------------------------------------+\n",
      "\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "| snp2_transcript_consequences.polyphen_score | snp2_transcript_consequences.protein_end |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "|                                     float64 |                                    int32 |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "|                                    8.94e-01 |                                        4 |\n",
      "|                                          NA |                                      487 |\n",
      "+---------------------------------------------+------------------------------------------+\n",
      "\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "| snp2_transcript_consequences.protein_start | snp2_transcript_consequences.protein_id |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                      int32 | str                                     |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                          4 | \"ENSP00000368720\"                       |\n",
      "|                                        487 | \"ENSP00000368720\"                       |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| snp2_transcript_consequences.sift_prediction | snp2_transcript_consequences.sift_score |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| str                                          |                                 float64 |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "| \"deleterious\"                                |                                3.00e-02 |\n",
      "| NA                                           |                                      NA |\n",
      "+----------------------------------------------+-----------------------------------------+\n",
      "\n",
      "+-------------------------------------+----------------------------------------+\n",
      "| snp2_transcript_consequences.strand | snp2_transcript_consequences.swissprot |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "|                               int32 | str                                    |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "|                                   1 | \"Q494U1\"                               |\n",
      "|                                   1 | \"Q494U1\"                               |\n",
      "+-------------------------------------+----------------------------------------+\n",
      "\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| snp2_transcript_consequences.transcript_id | snp2_transcript_consequences.trembl |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| str                                        | str                                 |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "| \"ENST00000379410\"                          | \"J3KSM5\"                            |\n",
      "| \"ENST00000379410\"                          | \"J3KSM5\"                            |\n",
      "+--------------------------------------------+-------------------------------------+\n",
      "\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| snp2_transcript_consequences.uniparc | snp2_transcript_consequences.variant_allele |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| str                                  | str                                         |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "| \"UPI00001416D8\"                      | \"A\"                                         |\n",
      "| \"UPI00001416D8\"                      | \"C\"                                         |\n",
      "+--------------------------------------+---------------------------------------------+\n",
      "\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| mnv_transcript_consequences.allele_num | mnv_transcript_consequences.amino_acids |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "|                                  int32 | str                                     |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "|                                      1 | \"S/K\"                                   |\n",
      "|                                      1 | \"R/P\"                                   |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "\n",
      "+-------------------------------------+---------------------------------------+\n",
      "| mnv_transcript_consequences.biotype | mnv_transcript_consequences.canonical |\n",
      "+-------------------------------------+---------------------------------------+\n",
      "| str                                 |                                 int32 |\n",
      "+-------------------------------------+---------------------------------------+\n",
      "| \"protein_coding\"                    |                                     1 |\n",
      "| \"protein_coding\"                    |                                     1 |\n",
      "+-------------------------------------+---------------------------------------+\n",
      "\n",
      "+----------------------------------+----------------------------------------+\n",
      "| mnv_transcript_consequences.ccds | mnv_transcript_consequences.cdna_start |\n",
      "+----------------------------------+----------------------------------------+\n",
      "| str                              |                                  int32 |\n",
      "+----------------------------------+----------------------------------------+\n",
      "| \"CCDS4.1\"                        |                                     46 |\n",
      "| \"CCDS4.1\"                        |                                   1495 |\n",
      "+----------------------------------+----------------------------------------+\n",
      "\n",
      "+--------------------------------------+-------------------------------------+\n",
      "| mnv_transcript_consequences.cdna_end | mnv_transcript_consequences.cds_end |\n",
      "+--------------------------------------+-------------------------------------+\n",
      "|                                int32 |                               int32 |\n",
      "+--------------------------------------+-------------------------------------+\n",
      "|                                   47 |                                  12 |\n",
      "|                                 1496 |                                1461 |\n",
      "+--------------------------------------+-------------------------------------+\n",
      "\n",
      "+---------------------------------------+------------------------------------+\n",
      "| mnv_transcript_consequences.cds_start | mnv_transcript_consequences.codons |\n",
      "+---------------------------------------+------------------------------------+\n",
      "|                                 int32 | str                                |\n",
      "+---------------------------------------+------------------------------------+\n",
      "|                                    11 | \"aGC/aAA\"                          |\n",
      "|                                  1460 | \"cGT/cCC\"                          |\n",
      "+---------------------------------------+------------------------------------+\n",
      "\n",
      "+-----------------------------------------------+--------------------------------------+\n",
      "| mnv_transcript_consequences.consequence_terms | mnv_transcript_consequences.distance |\n",
      "+-----------------------------------------------+--------------------------------------+\n",
      "| array<str>                                    |                                int32 |\n",
      "+-----------------------------------------------+--------------------------------------+\n",
      "| [\"missense_variant\"]                          |                                   NA |\n",
      "| [\"missense_variant\"]                          |                                   NA |\n",
      "+-----------------------------------------------+--------------------------------------+\n",
      "\n",
      "+---------------------------------------------------------+\n",
      "| mnv_transcript_consequences.domains                     |\n",
      "+---------------------------------------------------------+\n",
      "| array<struct{db: str, name: str}>                       |\n",
      "+---------------------------------------------------------+\n",
      "| NA                                                      |\n",
      "| [(\"hmmpanther\",\"PTHR22826\"),(\"hmmpanther\",\"PTHR22826\")] |\n",
      "+---------------------------------------------------------+\n",
      "\n",
      "+----------------------------------+-------------------------------------+\n",
      "| mnv_transcript_consequences.exon | mnv_transcript_consequences.gene_id |\n",
      "+----------------------------------+-------------------------------------+\n",
      "| str                              | str                                 |\n",
      "+----------------------------------+-------------------------------------+\n",
      "| \"1/16\"                           | \"ENSG00000187583\"                   |\n",
      "| \"14/16\"                          | \"ENSG00000187583\"                   |\n",
      "+----------------------------------+-------------------------------------+\n",
      "\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "| mnv_transcript_consequences.gene_pheno | mnv_transcript_consequences.gene_symbol |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "|                                  int32 | str                                     |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "|                                     NA | \"PLEKHN1\"                               |\n",
      "|                                     NA | \"PLEKHN1\"                               |\n",
      "+----------------------------------------+-----------------------------------------+\n",
      "\n",
      "+------------------------------------------------+-------------------------------------+\n",
      "| mnv_transcript_consequences.gene_symbol_source | mnv_transcript_consequences.hgnc_id |\n",
      "+------------------------------------------------+-------------------------------------+\n",
      "| str                                            | str                                 |\n",
      "+------------------------------------------------+-------------------------------------+\n",
      "| \"HGNC\"                                         | \"25284\"                             |\n",
      "| \"HGNC\"                                         | \"25284\"                             |\n",
      "+------------------------------------------------+-------------------------------------+\n",
      "\n",
      "+-------------------------------------------+-----------------------------------+\n",
      "| mnv_transcript_consequences.hgvsc         | mnv_transcript_consequences.hgvsp |\n",
      "+-------------------------------------------+-----------------------------------+\n",
      "| str                                       | str                               |\n",
      "+-------------------------------------------+-----------------------------------+\n",
      "| \"ENST00000379410.3:c.11_12delGCinsAA\"     | \"ENSP00000368720.3:p.Ser4Lys\"     |\n",
      "| \"ENST00000379410.3:c.1460_1461delGTinsCC\" | \"ENSP00000368720.3:p.Arg487Pro\"   |\n",
      "+-------------------------------------------+-----------------------------------+\n",
      "\n",
      "+-----------------------------------------+------------------------------------+\n",
      "| mnv_transcript_consequences.hgvs_offset | mnv_transcript_consequences.impact |\n",
      "+-----------------------------------------+------------------------------------+\n",
      "|                                   int32 | str                                |\n",
      "+-----------------------------------------+------------------------------------+\n",
      "|                                      NA | \"MODERATE\"                         |\n",
      "|                                      NA | \"MODERATE\"                         |\n",
      "+-----------------------------------------+------------------------------------+\n",
      "\n",
      "+------------------------------------+---------------------------------+\n",
      "| mnv_transcript_consequences.intron | mnv_transcript_consequences.lof |\n",
      "+------------------------------------+---------------------------------+\n",
      "| str                                | str                             |\n",
      "+------------------------------------+---------------------------------+\n",
      "| NA                                 | NA                              |\n",
      "| NA                                 | NA                              |\n",
      "+------------------------------------+---------------------------------+\n",
      "\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| mnv_transcript_consequences.lof_flags | mnv_transcript_consequences.lof_filter |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| str                                   | str                                    |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "| NA                                    | NA                                     |\n",
      "| NA                                    | NA                                     |\n",
      "+---------------------------------------+----------------------------------------+\n",
      "\n",
      "+--------------------------------------+---------------------------------------+\n",
      "| mnv_transcript_consequences.lof_info | mnv_transcript_consequences.minimised |\n",
      "+--------------------------------------+---------------------------------------+\n",
      "| str                                  |                                 int32 |\n",
      "+--------------------------------------+---------------------------------------+\n",
      "| NA                                   |                                     1 |\n",
      "| NA                                   |                                     1 |\n",
      "+--------------------------------------+---------------------------------------+\n",
      "\n",
      "+-------------------------------------------------+\n",
      "| mnv_transcript_consequences.polyphen_prediction |\n",
      "+-------------------------------------------------+\n",
      "| str                                             |\n",
      "+-------------------------------------------------+\n",
      "| \"possibly_damaging\"                             |\n",
      "| \"benign\"                                        |\n",
      "+-------------------------------------------------+\n",
      "\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "| mnv_transcript_consequences.polyphen_score | mnv_transcript_consequences.protein_end |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                    float64 |                                   int32 |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "|                                   7.51e-01 |                                       4 |\n",
      "|                                   0.00e+00 |                                     487 |\n",
      "+--------------------------------------------+-----------------------------------------+\n",
      "\n",
      "+-------------------------------------------+----------------------------------------+\n",
      "| mnv_transcript_consequences.protein_start | mnv_transcript_consequences.protein_id |\n",
      "+-------------------------------------------+----------------------------------------+\n",
      "|                                     int32 | str                                    |\n",
      "+-------------------------------------------+----------------------------------------+\n",
      "|                                         4 | \"ENSP00000368720\"                      |\n",
      "|                                       487 | \"ENSP00000368720\"                      |\n",
      "+-------------------------------------------+----------------------------------------+\n",
      "\n",
      "+---------------------------------------------+----------------------------------------+\n",
      "| mnv_transcript_consequences.sift_prediction | mnv_transcript_consequences.sift_score |\n",
      "+---------------------------------------------+----------------------------------------+\n",
      "| str                                         |                                float64 |\n",
      "+---------------------------------------------+----------------------------------------+\n",
      "| \"deleterious\"                               |                               5.00e-02 |\n",
      "| \"tolerated\"                                 |                               3.10e-01 |\n",
      "+---------------------------------------------+----------------------------------------+\n",
      "\n",
      "+------------------------------------+---------------------------------------+\n",
      "| mnv_transcript_consequences.strand | mnv_transcript_consequences.swissprot |\n",
      "+------------------------------------+---------------------------------------+\n",
      "|                              int32 | str                                   |\n",
      "+------------------------------------+---------------------------------------+\n",
      "|                                  1 | \"Q494U1\"                              |\n",
      "|                                  1 | \"Q494U1\"                              |\n",
      "+------------------------------------+---------------------------------------+\n",
      "\n",
      "+-------------------------------------------+------------------------------------+\n",
      "| mnv_transcript_consequences.transcript_id | mnv_transcript_consequences.trembl |\n",
      "+-------------------------------------------+------------------------------------+\n",
      "| str                                       | str                                |\n",
      "+-------------------------------------------+------------------------------------+\n",
      "| \"ENST00000379410\"                         | \"J3KSM5\"                           |\n",
      "| \"ENST00000379410\"                         | \"J3KSM5\"                           |\n",
      "+-------------------------------------------+------------------------------------+\n",
      "\n",
      "+-------------------------------------+--------------------------------------------+\n",
      "| mnv_transcript_consequences.uniparc | mnv_transcript_consequences.variant_allele |\n",
      "+-------------------------------------+--------------------------------------------+\n",
      "| str                                 | str                                        |\n",
      "+-------------------------------------+--------------------------------------------+\n",
      "| \"UPI00001416D8\"                     | \"AA\"                                       |\n",
      "| \"UPI00001416D8\"                     | \"CC\"                                       |\n",
      "+-------------------------------------+--------------------------------------------+\n",
      "\n",
      "+----------------------+------------------------+----------------------+-------------+\n",
      "| snp1_cons_term       | snp2_cons_term         | mnv_cons_term        | snp1_codons |\n",
      "+----------------------+------------------------+----------------------+-------------+\n",
      "| array<str>           | array<str>             | array<str>           | str         |\n",
      "+----------------------+------------------------+----------------------+-------------+\n",
      "| [\"missense_variant\"] | [\"missense_variant\"]   | [\"missense_variant\"] | \"aGc/aAc\"   |\n",
      "| [\"missense_variant\"] | [\"synonymous_variant\"] | [\"missense_variant\"] | \"cGt/cCt\"   |\n",
      "+----------------------+------------------------+----------------------+-------------+\n",
      "\n",
      "+-------------+------------+------------------+------------------+-----------------+\n",
      "| snp2_codons | mnv_codons | snp1_amino_acids | snp2_amino_acids | mnv_amino_acids |\n",
      "+-------------+------------+------------------+------------------+-----------------+\n",
      "| str         | str        | str              | str              | str             |\n",
      "+-------------+------------+------------------+------------------+-----------------+\n",
      "| \"agC/agA\"   | \"aGC/aAA\"  | \"S/N\"            | \"S/R\"            | \"S/K\"           |\n",
      "| \"cgT/cgC\"   | \"cGT/cCC\"  | \"R/P\"            | \"R\"              | \"R/P\"           |\n",
      "+-------------+------------+------------------+------------------+-----------------+\n",
      "\n",
      "+----------+----------+---------+-------------------+\n",
      "| snp1_lof | snp2_lof | mnv_lof | transcript_id     |\n",
      "+----------+----------+---------+-------------------+\n",
      "| str      | str      | str     | str               |\n",
      "+----------+----------+---------+-------------------+\n",
      "| NA       | NA       | NA      | \"ENST00000379410\" |\n",
      "| NA       | NA       | NA      | \"ENST00000379410\" |\n",
      "+----------+----------+---------+-------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "canon_cons_d1.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the last step, we will actually use traditional dataframe approach, by turning the table to pandas dataframe.\n",
    "(the flexibility of `apply` command pandas has is still attractive!)\n",
    "\n",
    "Then we can annotate the most severe consequence out of all the possible consequences, \n",
    "and also categorize the MNV by comparing the consequence of SNP1, SNP2, and MNV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-03-10 01:50:09 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:50:09 Hail: INFO: Coerced sorted dataset\n",
      "2019-03-10 01:50:10 Hail: INFO: Coerced sorted dataset\n"
     ]
    }
   ],
   "source": [
    "#keep the necessary columns and turn to pandas dataframe\n",
    "canon_cons_pd1 = canon_cons_d1.select(\"snp1_cons_term\", \"snp2_cons_term\", \"mnv_cons_term\", \"snp1_codons\",\n",
    "                                              \"snp2_codons\", \"mnv_codons\", \"snp1_amino_acids\", \"snp2_amino_acids\",\n",
    "                                              \"mnv_amino_acids\", \"snp1_lof\", \"snp2_lof\", \"mnv_lof\", \"transcript_id\",\n",
    "                                              \"AC\", \"prev_AC\", \"n_hethet\", \"n_hethom\", \"n_homhom\").to_pandas()\n",
    "# get the most severe consequence\n",
    "canon_cons_pd1[\"snp1_sev\"] = canon_cons_pd1.snp1_cons_term.apply(lambda x: cons_term_most_severe(x))\n",
    "canon_cons_pd1[\"snp2_sev\"] = canon_cons_pd1.snp2_cons_term.apply(lambda x: cons_term_most_severe(x))\n",
    "canon_cons_pd1[\"mnv_sev\"] = canon_cons_pd1.mnv_cons_term.apply(lambda x: cons_term_most_severe(x))\n",
    "# annotate the category\n",
    "canon_cons_pd1[\"categ\"] = canon_cons_pd1.apply(\n",
    "            lambda x: mnv_category(x[\"snp1_sev\"], x[\"snp2_sev\"], x[\"mnv_sev\"], x[\"snp1_amino_acids\"],\n",
    "                                   x[\"snp2_amino_acids\"],\n",
    "                                   x[\"mnv_amino_acids\"]), axis=1)\n",
    "#rewrite lof flags just in case if lof column is all None (NA) and that causes error\n",
    "canon_cons_pd1.snp1_lof = canon_cons_pd1.snp1_lof.astype(str)\n",
    "canon_cons_pd1.snp2_lof = canon_cons_pd1.snp2_lof.astype(str)\n",
    "canon_cons_pd1.mnv_lof = canon_cons_pd1.mnv_lof.astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus.contig</th>\n",
       "      <th>locus.position</th>\n",
       "      <th>refs</th>\n",
       "      <th>alts</th>\n",
       "      <th>snp1_cons_term</th>\n",
       "      <th>snp2_cons_term</th>\n",
       "      <th>mnv_cons_term</th>\n",
       "      <th>snp1_codons</th>\n",
       "      <th>snp2_codons</th>\n",
       "      <th>mnv_codons</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>AC</th>\n",
       "      <th>prev_AC</th>\n",
       "      <th>n_hethet</th>\n",
       "      <th>n_hethom</th>\n",
       "      <th>n_homhom</th>\n",
       "      <th>snp1_sev</th>\n",
       "      <th>snp2_sev</th>\n",
       "      <th>mnv_sev</th>\n",
       "      <th>categ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>901922</td>\n",
       "      <td>GC</td>\n",
       "      <td>AA</td>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>aGc/aAc</td>\n",
       "      <td>agC/agA</td>\n",
       "      <td>aGC/aAA</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379410</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Changed missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>909238</td>\n",
       "      <td>GT</td>\n",
       "      <td>CC</td>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>[missense_variant]</td>\n",
       "      <td>cGt/cCt</td>\n",
       "      <td>cgT/cgC</td>\n",
       "      <td>cGT/cCC</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000379410</td>\n",
       "      <td>1</td>\n",
       "      <td>3889</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Unchanged</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus.contig  locus.position refs alts      snp1_cons_term  \\\n",
       "0            1          901922   GC   AA  [missense_variant]   \n",
       "1            1          909238   GT   CC  [missense_variant]   \n",
       "\n",
       "         snp2_cons_term       mnv_cons_term snp1_codons snp2_codons  \\\n",
       "0    [missense_variant]  [missense_variant]     aGc/aAc     agC/agA   \n",
       "1  [synonymous_variant]  [missense_variant]     cGt/cCt     cgT/cgC   \n",
       "\n",
       "  mnv_codons  ...    transcript_id  AC prev_AC n_hethet n_hethom n_homhom  \\\n",
       "0    aGC/aAA  ...  ENST00000379410  53      53       51        1        0   \n",
       "1    cGT/cCC  ...  ENST00000379410   1    3889        0        0        1   \n",
       "\n",
       "           snp1_sev            snp2_sev           mnv_sev             categ  \n",
       "0  missense_variant    missense_variant  missense_variant  Changed missense  \n",
       "1  missense_variant  synonymous_variant  missense_variant         Unchanged  \n",
       "\n",
       "[2 rows x 26 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(canon_cons_pd1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## End notes\n",
    "\n",
    "- After all, it turns out that there are only 2 MNVs of distance 1 that falls within a codon reading frame. One of them is \"Changed missense\", and the other is \"Unchanged\".\n",
    "\n",
    "(Not surprising considering that we were just playing around with a small chunk of region -- again the very useful case of cloud computation is when dealing with larger dataset.)\n",
    "\n",
    "- We can do pretty much the same thing in MNVs of distance 2 (just replace the `vepped_d1_essense = t1.key_by(\"locus\", \"refs\", \"alts\") ` with \n",
    "`vepped_d2_essense = t2.key_by(\"locus\", \"refs\", \"alts\") ` and all the downstream `d1` to `d2` for clarification.).\n",
    "\n",
    "- (also note that running vep after annotating the sequence context sometimes raise, because of memory issue. In such case we recommend to increase the number of clusters.)\n",
    "\n",
    "- Again we do provide a function `annotate_vep_mnv` that does all the processes above in a single line. Note that we need to specify the distance of the MNV for this function (i.e. you will run this function 2 times, one for distance 1, and another for distance 2)\n",
    "\n",
    "- In the next section `functional_impact.ipynb`, we will use the real summary data from `gnomAD` exome -- a collection of 31,510 MNVs falling within a single codon!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions\n",
    "Below are the list of functions we defined in this notebook (before running all the processes above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_vep_to_canonical_transcripts(mt: Union[hl.MatrixTable, hl.Table],\n",
    "                                        vep_root: str = 'vep') -> Union[hl.MatrixTable, hl.Table]:\n",
    "    canonical = mt[vep_root].transcript_consequences.filter(lambda csq: csq.canonical == 1)\n",
    "    vep_data = mt[vep_root].annotate(transcript_consequences=canonical)\n",
    "    return mt.annotate_rows(**{vep_root: vep_data}) if isinstance(mt, hl.MatrixTable) else mt.annotate(**{vep_root: vep_data})\n",
    "\n",
    "def mnv_category_by_aa_change(snp1_con, snp2_con, mnv_con,aa1,aa2,aa3):\n",
    "    if (snp1_con, snp2_con, mnv_con)==(\"synonymous_variant\",\"missense_variant\",\"missense_variant\"):\n",
    "        if aa2==aa3: return (\"Unchanged\")\n",
    "        else: return (\"Changed missense\")\n",
    "    elif (snp1_con, snp2_con, mnv_con)==(\"missense_variant\",\"synonymous_variant\",\"missense_variant\"):\n",
    "        if aa1==aa3: return (\"Unchanged\")\n",
    "        else: return (\"Changed missense\")\n",
    "    elif (snp1_con, snp2_con, mnv_con)==(\"missense_variant\",\"missense_variant\",\"missense_variant\"):\n",
    "        if ((aa1==aa3) or (aa2==aa3)):\n",
    "            return (\"Partially changed missense\")\n",
    "        else: return (\"Changed missense\")\n",
    "    else: return (\"something wrong going on\")\n",
    "\n",
    "def cons_term_most_severe(cons_term_array):\n",
    "    if \"start_lost\" in cons_term_array: return (\"start_lost\") #by definition this is sufficient to determine the mnv is uninteresting\n",
    "    elif \"stop_lost\" in cons_term_array: return (\"stop_lost\")\n",
    "    elif \"stop_gained\" in cons_term_array: return (\"stop_gained\")\n",
    "    elif \"missense_variant\" in cons_term_array: return (\"missense_variant\")\n",
    "    elif \"stop_retained_variant\" in cons_term_array: return (\"stop_retained_variant\")\n",
    "    elif \"synonymous_variant\" in cons_term_array: return (\"synonymous_variant\")\n",
    "    else: return (\"Noncoding_or_else\")\n",
    "\n",
    "def mnv_category(snp1_con, snp2_con, mnv_con, aa1, aa2, aa3):\n",
    "    # return the MNV consequence category, such as gained PTV, unchange, etc.\n",
    "    # just classify everything of 3*3*3=27 pattern.\n",
    "    #plus, case where we see start_lost / stop_lost\n",
    "    # and if undeterminisitic, look at the aa change and determine.\n",
    "    if snp1_con == \"synonymous_variant\":\n",
    "        if snp2_con == \"synonymous_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Unchanged\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Gained missense\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Gained PTV\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"missense_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Lost missense\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (mnv_category_by_aa_change(snp1_con, snp2_con, mnv_con, aa1, aa2, aa3))  # go look at aa change.\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Gained PTV\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"stop_gained\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Unchanged\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "    if snp1_con == \"missense_variant\":\n",
    "        if snp2_con == \"synonymous_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Lost missense\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (mnv_category_by_aa_change(snp1_con, snp2_con, mnv_con, aa1, aa2, aa3))  # go look at aa change.\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Gained PTV\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"missense_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Lost missense\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (mnv_category_by_aa_change(snp1_con, snp2_con, mnv_con, aa1, aa2, aa3))  # go look at aa change.\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Gained PTV\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"stop_gained\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Unchanged\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        else:\n",
    "            return (\"Noncoding_or_else\")\n",
    "    if snp1_con == \"stop_gained\":\n",
    "        if snp2_con == \"synonymous_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Unchanged\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"missense_variant\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Unchanged\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        if snp2_con == \"stop_gained\":\n",
    "            if mnv_con == \"synonymous_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"missense_variant\":\n",
    "                return (\"Rescued PTV\")\n",
    "            elif mnv_con == \"stop_gained\":\n",
    "                return (\"Unchanged\")\n",
    "            else:\n",
    "                return (\"Noncoding_or_else\")\n",
    "        else:\n",
    "            return (\"Noncoding_or_else\")\n",
    "    #else, involving start_loss etc -> look at mnv cons first.\n",
    "    elif mnv_con==\"start_lost\": return \"Unchanged\" #by definition individual effect is also start loss\n",
    "    elif mnv_con==\"stop_lost\":\n",
    "        if ((snp1_con==\"stop_retained_variant\") & (snp2_con==\"stop_retained_variant\")): return \"gained_stop_loss\"\n",
    "        else: return (\"Unchanged\")\n",
    "    elif mnv_con==\"stop_retained_variant\":#this case, by definition one of the variant is stop_lost, and the other is stop_retained\n",
    "        return (\"Rescued stop loss\")\n",
    "    else:\n",
    "        return (\"Noncoding_or_else\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Hail",
   "language": "python",
   "name": "hail"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
